{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2375f9e6-dbd7-4acd-8f42-fda408e04b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import h5py\n",
    "import pandas as pd\n",
    "from itertools import islice\n",
    "from collections import Counter\n",
    "import re\n",
    "from itertools import islice\n",
    "from collections import Counter\n",
    "from io import BytesIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b53bc8b2-6bc9-497c-a211-344ee3bd803d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data\n",
      "meta\n",
      "genes\n",
      "info\n",
      "samples\n",
      "<KeysViewHDF5 ['chromosome', 'ensembl_gene_id', 'gene_biotype', 'gene_symbol', 'genes', 'start_position']>\n"
     ]
    }
   ],
   "source": [
    "mouse = h5py.File('mouse_matrix_v11.h5', \"r\") # fot reading h5 dataframe\n",
    "data=mouse.keys()\n",
    "for key in mouse.keys():\n",
    "    print(key)\n",
    "for key in mouse['meta'].keys():\n",
    "    print(key)\n",
    "list_of_columns=mouse['meta']['genes'].keys()\n",
    "print(list_of_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dbf0f137-34f6-4594-bf69-46ecdfd5a6ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome\n",
      "[b'missing' b'11' b'11' ... b'MT' b'MT' b'MT']\n",
      "ensembl_gene_id\n",
      "[b'missing' b'ENSMUSG00000007777' b'ENSMUSG00000043644' ...\n",
      " b'ENSMUSG00000065947' b'ENSMUSG00000064367' b'ENSMUSG00000064368']\n",
      "gene_biotype\n",
      "[b'missing' b'protein_coding' b'lncRNA' ... b'protein_coding'\n",
      " b'protein_coding' b'protein_coding']\n",
      "gene_symbol\n",
      "[b'0610007P14Rik' b'0610009B22Rik' b'0610009L18Rik' ... b'mt-Nd4l'\n",
      " b'mt-Nd5' b'mt-Nd6']\n",
      "genes\n",
      "[b'0610007P14Rik' b'0610009B22Rik' b'0610009L18Rik' ... b'mt-Nd4l'\n",
      " b'mt-Nd5' b'mt-Nd6']\n",
      "start_position\n",
      "[b'missing' b'51576213' b'120239504' ... b'9877' b'11742' b'13552']\n"
     ]
    }
   ],
   "source": [
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "    for key in f[\"meta/genes\"].keys():\n",
    "        a = f[\"meta/genes/\"][key]\n",
    "        print(key)\n",
    "        print(a[:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73b04b0d-9ca6-4e4d-bd88-dd5c93778401",
   "metadata": {},
   "source": [
    "Now every dataset from h5 file we transform into single table in pandas and then concatenate them into one table\n",
    "import h5py\n",
    "arch_file = \"/home/clever/test_cache/counts/archs4/mouse_matrix_v11.h5\"\n",
    "with h5py.File(arch_file, 'r') as arch_7:\n",
    "    for key in arch_7[\"meta/samples\"].keys():\n",
    "        a = arch_7[\"meta/samples/\"][key]\n",
    "        \n",
    "        print(key)\n",
    "        print(a[:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "42a49fa5-7619-4914-8c0f-fefe47eee332",
   "metadata": {},
   "outputs": [],
   "source": [
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "    data_set=f['meta']['samples']\n",
    "    lister=['taxid_ch1','status','source_name_ch1','series_id','relation','readstotal','readsaligned','platform_id','organism_ch1', 'molecule_ch1','library_strategy','library_selection', 'library_source','instrument_model','extract_protocol_ch1', 'characteristics_ch1','data_processing','geo_accession','type','title']\n",
    "    for i in lister:\n",
    "        a=data_set[f'{i}']\n",
    "        a=pd.DataFrame(a, columns=[f'{i}'])\n",
    "        for col, dtype in a.dtypes.items():\n",
    "            if dtype == np.object_:  # Only process byte object columns.\n",
    "                a[col] = a[col].apply(lambda x: x.decode(\"utf-8\"))\n",
    "        table=pd.concat([table, a],sort=False, axis=1)\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "fcf65c60-4d22-4c61-a0e1-7054d274bbf5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>status</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>series_id</th>\n",
       "      <th>relation</th>\n",
       "      <th>readstotal</th>\n",
       "      <th>readsaligned</th>\n",
       "      <th>platform_id</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>molecule_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_source</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>extract_protocol_ch1</th>\n",
       "      <th>characteristics_ch1</th>\n",
       "      <th>data_processing</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>type</th>\n",
       "      <th>title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Apr 07 2015'</td>\n",
       "      <td>b'kidney cortex'</td>\n",
       "      <td>b'GSE54390'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>39116064</td>\n",
       "      <td>32884497</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'total RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Thin outer slice of kidney cortex was immediat...</td>\n",
       "      <td>tissue: kidney cortex\\tstrain: C57BL/6\\tgender...</td>\n",
       "      <td>Reads were aligned to the genome using the Bur...</td>\n",
       "      <td>GSM1314260</td>\n",
       "      <td>SRA</td>\n",
       "      <td>TRACK 43-75 kidney 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'corpus callosum'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1136541</td>\n",
       "      <td>308433</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'polyA RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2164318</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772117-029-A02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'dorsal horn'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>514285</td>\n",
       "      <td>148668</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'polyA RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2161048</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772072-242-F03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jan 01 2016'</td>\n",
       "      <td>b'Lung tissue'</td>\n",
       "      <td>b'GSE64027'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>26804780</td>\n",
       "      <td>23355974</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'total RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was extracted from pooled dissected ...</td>\n",
       "      <td>strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...</td>\n",
       "      <td>Illumina Casava1.8 software used for basecalli...</td>\n",
       "      <td>GSM1563082</td>\n",
       "      <td>SRA</td>\n",
       "      <td>20NNK_LPS_Sil_I3C9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2012'</td>\n",
       "      <td>b'Cortex'</td>\n",
       "      <td>b'GSE36025\\tGSE49417\\tGSE49847'</td>\n",
       "      <td>b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...</td>\n",
       "      <td>265978941</td>\n",
       "      <td>173619445</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'polyA RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...</td>\n",
       "      <td>lab: CSHL-m\\tlab description: Gingeras - Cold ...</td>\n",
       "      <td>http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...</td>\n",
       "      <td>GSM1000563</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501049</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shGAA lentivirus'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>13275033</td>\n",
       "      <td>12476567</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'total RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592485</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_GAA 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501050</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shscrambled'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>12754218</td>\n",
       "      <td>11967090</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'total RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592486</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_SC 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501051</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shscrambled'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>12243444</td>\n",
       "      <td>11535137</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'total RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592487</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_SC 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501052</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shscrambled'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>12163822</td>\n",
       "      <td>11405087</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'total RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592488</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_SC 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501053</th>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Feb 06 2017'</td>\n",
       "      <td>b'BCR-ABL transformed in vitro cell culture'</td>\n",
       "      <td>b'GSE39160'</td>\n",
       "      <td>b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...</td>\n",
       "      <td>145311674</td>\n",
       "      <td>87715899</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>b'polyA RNA'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>polyA-enriched RNA was purified with the Micro...</td>\n",
       "      <td>strain: C57BL/6\\tgenotype/variation: Ikaros_Zn...</td>\n",
       "      <td>Raw data (fastq) was uploaded to the Galaxy we...</td>\n",
       "      <td>GSM957182</td>\n",
       "      <td>SRA</td>\n",
       "      <td>BCR-ABL+_Ikaros_ZnF4-/-_d14_repl1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>501054 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       taxid_ch1                    status  \\\n",
       "0       b'10090'  b'Public on Apr 07 2015'   \n",
       "1       b'10090'  b'Public on Jun 10 2016'   \n",
       "2       b'10090'  b'Public on Jun 10 2016'   \n",
       "3       b'10090'  b'Public on Jan 01 2016'   \n",
       "4       b'10090'  b'Public on Sep 10 2012'   \n",
       "...          ...                       ...   \n",
       "501049  b'10090'  b'Public on Sep 23 2021'   \n",
       "501050  b'10090'  b'Public on Sep 23 2021'   \n",
       "501051  b'10090'  b'Public on Sep 23 2021'   \n",
       "501052  b'10090'  b'Public on Sep 23 2021'   \n",
       "501053  b'10090'  b'Public on Feb 06 2017'   \n",
       "\n",
       "                                     source_name_ch1  \\\n",
       "0                                   b'kidney cortex'   \n",
       "1                                 b'corpus callosum'   \n",
       "2                                     b'dorsal horn'   \n",
       "3                                     b'Lung tissue'   \n",
       "4                                          b'Cortex'   \n",
       "...                                              ...   \n",
       "501049          b'transfected with shGAA lentivirus'   \n",
       "501050               b'transfected with shscrambled'   \n",
       "501051               b'transfected with shscrambled'   \n",
       "501052               b'transfected with shscrambled'   \n",
       "501053  b'BCR-ABL transformed in vitro cell culture'   \n",
       "\n",
       "                              series_id  \\\n",
       "0                           b'GSE54390'   \n",
       "1                           b'GSE75330'   \n",
       "2                           b'GSE75330'   \n",
       "3                           b'GSE64027'   \n",
       "4       b'GSE36025\\tGSE49417\\tGSE49847'   \n",
       "...                                 ...   \n",
       "501049                     b'GSE184555'   \n",
       "501050                     b'GSE184555'   \n",
       "501051                     b'GSE184555'   \n",
       "501052                     b'GSE184555'   \n",
       "501053                      b'GSE39160'   \n",
       "\n",
       "                                                 relation  readstotal  \\\n",
       "0       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    39116064   \n",
       "1       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1136541   \n",
       "2       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      514285   \n",
       "3       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    26804780   \n",
       "4       b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...   265978941   \n",
       "...                                                   ...         ...   \n",
       "501049  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    13275033   \n",
       "501050  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    12754218   \n",
       "501051  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    12243444   \n",
       "501052  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    12163822   \n",
       "501053  b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...   145311674   \n",
       "\n",
       "        readsaligned platform_id     organism_ch1  molecule_ch1  ...  \\\n",
       "0           32884497    GPL13112  b'Mus musculus'  b'total RNA'  ...   \n",
       "1             308433    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
       "2             148668    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
       "3           23355974    GPL13112  b'Mus musculus'  b'total RNA'  ...   \n",
       "4          173619445    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
       "...              ...         ...              ...           ...  ...   \n",
       "501049      12476567    GPL21626  b'Mus musculus'  b'total RNA'  ...   \n",
       "501050      11967090    GPL21626  b'Mus musculus'  b'total RNA'  ...   \n",
       "501051      11535137    GPL21626  b'Mus musculus'  b'total RNA'  ...   \n",
       "501052      11405087    GPL21626  b'Mus musculus'  b'total RNA'  ...   \n",
       "501053      87715899    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
       "\n",
       "       library_strategy library_selection  library_source  \\\n",
       "0               RNA-Seq              cDNA  transcriptomic   \n",
       "1               RNA-Seq              cDNA  transcriptomic   \n",
       "2               RNA-Seq              cDNA  transcriptomic   \n",
       "3               RNA-Seq              cDNA  transcriptomic   \n",
       "4               RNA-Seq              cDNA  transcriptomic   \n",
       "...                 ...               ...             ...   \n",
       "501049          RNA-Seq              cDNA  transcriptomic   \n",
       "501050          RNA-Seq              cDNA  transcriptomic   \n",
       "501051          RNA-Seq              cDNA  transcriptomic   \n",
       "501052          RNA-Seq              cDNA  transcriptomic   \n",
       "501053          RNA-Seq              cDNA  transcriptomic   \n",
       "\n",
       "           instrument_model  \\\n",
       "0       Illumina HiSeq 2000   \n",
       "1       Illumina HiSeq 2000   \n",
       "2       Illumina HiSeq 2000   \n",
       "3       Illumina HiSeq 2000   \n",
       "4       Illumina HiSeq 2000   \n",
       "...                     ...   \n",
       "501049          NextSeq 550   \n",
       "501050          NextSeq 550   \n",
       "501051          NextSeq 550   \n",
       "501052          NextSeq 550   \n",
       "501053  Illumina HiSeq 2000   \n",
       "\n",
       "                                     extract_protocol_ch1  \\\n",
       "0       Thin outer slice of kidney cortex was immediat...   \n",
       "1       Selected CNS regions were dissected in 300µm t...   \n",
       "2       Selected CNS regions were dissected in 300µm t...   \n",
       "3       Total RNA was extracted from pooled dissected ...   \n",
       "4       http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...   \n",
       "...                                                   ...   \n",
       "501049  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501050  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501051  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501052  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501053  polyA-enriched RNA was purified with the Micro...   \n",
       "\n",
       "                                      characteristics_ch1  \\\n",
       "0       tissue: kidney cortex\\tstrain: C57BL/6\\tgender...   \n",
       "1       strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...   \n",
       "2       strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...   \n",
       "3       strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...   \n",
       "4       lab: CSHL-m\\tlab description: Gingeras - Cold ...   \n",
       "...                                                   ...   \n",
       "501049  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501050  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501051  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501052  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501053  strain: C57BL/6\\tgenotype/variation: Ikaros_Zn...   \n",
       "\n",
       "                                          data_processing geo_accession type  \\\n",
       "0       Reads were aligned to the genome using the Bur...    GSM1314260  SRA   \n",
       "1       Read processing was performed as described (Is...    GSM2164318  SRA   \n",
       "2       Read processing was performed as described (Is...    GSM2161048  SRA   \n",
       "3       Illumina Casava1.8 software used for basecalli...    GSM1563082  SRA   \n",
       "4       http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...    GSM1000563  SRA   \n",
       "...                                                   ...           ...  ...   \n",
       "501049  Raw sequence that passed CASAV 1.8P/F filter w...    GSM5592485  SRA   \n",
       "501050  Raw sequence that passed CASAV 1.8P/F filter w...    GSM5592486  SRA   \n",
       "501051  Raw sequence that passed CASAV 1.8P/F filter w...    GSM5592487  SRA   \n",
       "501052  Raw sequence that passed CASAV 1.8P/F filter w...    GSM5592488  SRA   \n",
       "501053  Raw data (fastq) was uploaded to the Galaxy we...     GSM957182  SRA   \n",
       "\n",
       "                                                    title  \n",
       "0                                    TRACK 43-75 kidney 1  \n",
       "1                                      C1-1772117-029-A02  \n",
       "2                                      C1-1772072-242-F03  \n",
       "3                                      20NNK_LPS_Sil_I3C9  \n",
       "4       CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...  \n",
       "...                                                   ...  \n",
       "501049                                          Fed_GAA 3  \n",
       "501050                                           Fed_SC 1  \n",
       "501051                                           Fed_SC 2  \n",
       "501052                                           Fed_SC 3  \n",
       "501053                  BCR-ABL+_Ikaros_ZnF4-/-_d14_repl1  \n",
       "\n",
       "[501054 rows x 40 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8ec6f4c3-e94e-40c3-b4de-9472d5eb57bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "table.to_csv('table.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "595ccb12-b894-4296-b839-4db10fc096fd",
   "metadata": {},
   "source": [
    "### Декодируем все строки из юникод в обычные (убираем b' ')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77fe03d6-75fa-4d65-91ed-39e9df740187",
   "metadata": {},
   "source": [
    "for col, dtype in table.dtypes.items():\n",
    "    if dtype == np.object_:  # Only process byte object columns.\n",
    "        table[col] = table[col].apply(lambda x: x.decode(\"utf-8\"))\n",
    "table.to_csv('table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fdcb4737-4ba1-4115-b33f-aaf627d468a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  taxid_ch1                    status     source_name_ch1  \\\n",
      "0  b'10090'  b'Public on Apr 07 2015'    b'kidney cortex'   \n",
      "1  b'10090'  b'Public on Jun 10 2016'  b'corpus callosum'   \n",
      "2  b'10090'  b'Public on Jun 10 2016'      b'dorsal horn'   \n",
      "3  b'10090'  b'Public on Jan 01 2016'      b'Lung tissue'   \n",
      "4  b'10090'  b'Public on Sep 10 2012'           b'Cortex'   \n",
      "\n",
      "                         series_id  \\\n",
      "0                      b'GSE54390'   \n",
      "1                      b'GSE75330'   \n",
      "2                      b'GSE75330'   \n",
      "3                      b'GSE64027'   \n",
      "4  b'GSE36025\\tGSE49417\\tGSE49847'   \n",
      "\n",
      "                                            relation  readstotal  \\\n",
      "0  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    39116064   \n",
      "1  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1136541   \n",
      "2  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      514285   \n",
      "3  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    26804780   \n",
      "4  b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...   265978941   \n",
      "\n",
      "   readsaligned platform_id     organism_ch1  molecule_ch1  ...  \\\n",
      "0      32884497    GPL13112  b'Mus musculus'  b'total RNA'  ...   \n",
      "1        308433    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
      "2        148668    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
      "3      23355974    GPL13112  b'Mus musculus'  b'total RNA'  ...   \n",
      "4     173619445    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
      "\n",
      "  library_strategy library_selection  library_source     instrument_model  \\\n",
      "0          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "1          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "2          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "3          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "4          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "\n",
      "                                extract_protocol_ch1  \\\n",
      "0  Thin outer slice of kidney cortex was immediat...   \n",
      "1  Selected CNS regions were dissected in 300µm t...   \n",
      "2  Selected CNS regions were dissected in 300µm t...   \n",
      "3  Total RNA was extracted from pooled dissected ...   \n",
      "4  http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...   \n",
      "\n",
      "                                 characteristics_ch1  \\\n",
      "0  tissue: kidney cortex\\tstrain: C57BL/6\\tgender...   \n",
      "1  strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...   \n",
      "2  strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...   \n",
      "3  strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...   \n",
      "4  lab: CSHL-m\\tlab description: Gingeras - Cold ...   \n",
      "\n",
      "                                     data_processing geo_accession type  \\\n",
      "0  Reads were aligned to the genome using the Bur...    GSM1314260  SRA   \n",
      "1  Read processing was performed as described (Is...    GSM2164318  SRA   \n",
      "2  Read processing was performed as described (Is...    GSM2161048  SRA   \n",
      "3  Illumina Casava1.8 software used for basecalli...    GSM1563082  SRA   \n",
      "4  http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...    GSM1000563  SRA   \n",
      "\n",
      "                                               title  \n",
      "0                               TRACK 43-75 kidney 1  \n",
      "1                                 C1-1772117-029-A02  \n",
      "2                                 C1-1772072-242-F03  \n",
      "3                                 20NNK_LPS_Sil_I3C9  \n",
      "4  CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...  \n",
      "\n",
      "[5 rows x 40 columns]\n"
     ]
    }
   ],
   "source": [
    "print(table.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1d853f2e-c637-4805-98d2-8032ae9d6b02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM1849922\n"
     ]
    }
   ],
   "source": [
    "print(geo_accession[28].decode(\"utf-8\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06d19b29-34ef-4c26-83d6-c142f16eb08e",
   "metadata": {},
   "source": [
    "Let's delete rows with words - single cell in a column 'source_name_ch1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56f27395-032f-41c3-ab13-8e3790441fe1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_1162/442747323.py:1: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('table.csv') # read table into pandas dataframe\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('table.csv') # read table into pandas dataframe\n",
    "discard=[\"single cell\"]\n",
    "del_single=df[~df.source_name_ch1.str.contains('|'.join(discard))] # ~ - все, кроме того что дальше - после нее\n",
    "del_single.to_csv('del_single.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "dc5bda68-8f52-4dd2-be41-15c358cff2b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>status</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>series_id</th>\n",
       "      <th>relation</th>\n",
       "      <th>readstotal</th>\n",
       "      <th>readsaligned</th>\n",
       "      <th>platform_id</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>library_strategy.1</th>\n",
       "      <th>library_selection.1</th>\n",
       "      <th>library_source.1</th>\n",
       "      <th>instrument_model.1</th>\n",
       "      <th>extract_protocol_ch1.1</th>\n",
       "      <th>characteristics_ch1.1</th>\n",
       "      <th>data_processing.1</th>\n",
       "      <th>geo_accession.1</th>\n",
       "      <th>type.1</th>\n",
       "      <th>title.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Apr 07 2015'</td>\n",
       "      <td>b'kidney cortex'</td>\n",
       "      <td>b'GSE54390'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>39116064</td>\n",
       "      <td>32884497</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Thin outer slice of kidney cortex was immediat...</td>\n",
       "      <td>tissue: kidney cortex\\tstrain: C57BL/6\\tgender...</td>\n",
       "      <td>Reads were aligned to the genome using the Bur...</td>\n",
       "      <td>GSM1314260</td>\n",
       "      <td>SRA</td>\n",
       "      <td>TRACK 43-75 kidney 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'corpus callosum'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1136541</td>\n",
       "      <td>308433</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2164318</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772117-029-A02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'dorsal horn'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>514285</td>\n",
       "      <td>148668</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2161048</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772072-242-F03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jan 01 2016'</td>\n",
       "      <td>b'Lung tissue'</td>\n",
       "      <td>b'GSE64027'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>26804780</td>\n",
       "      <td>23355974</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was extracted from pooled dissected ...</td>\n",
       "      <td>strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...</td>\n",
       "      <td>Illumina Casava1.8 software used for basecalli...</td>\n",
       "      <td>GSM1563082</td>\n",
       "      <td>SRA</td>\n",
       "      <td>20NNK_LPS_Sil_I3C9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2012'</td>\n",
       "      <td>b'Cortex'</td>\n",
       "      <td>b'GSE36025\\tGSE49417\\tGSE49847'</td>\n",
       "      <td>b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...</td>\n",
       "      <td>265978941</td>\n",
       "      <td>173619445</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...</td>\n",
       "      <td>lab: CSHL-m\\tlab description: Gingeras - Cold ...</td>\n",
       "      <td>http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...</td>\n",
       "      <td>GSM1000563</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501049</th>\n",
       "      <td>501049</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shGAA lentivirus'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>13275033</td>\n",
       "      <td>12476567</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592485</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_GAA 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501050</th>\n",
       "      <td>501050</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shscrambled'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>12754218</td>\n",
       "      <td>11967090</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592486</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_SC 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501051</th>\n",
       "      <td>501051</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shscrambled'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>12243444</td>\n",
       "      <td>11535137</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592487</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_SC 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501052</th>\n",
       "      <td>501052</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 23 2021'</td>\n",
       "      <td>b'transfected with shscrambled'</td>\n",
       "      <td>b'GSE184555'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>12163822</td>\n",
       "      <td>11405087</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>RNA was extracted using a QIAGEN Rneasy miniki...</td>\n",
       "      <td>cell line: C2C12\\tdifferentiation protocol: Di...</td>\n",
       "      <td>Raw sequence that passed CASAV 1.8P/F filter w...</td>\n",
       "      <td>GSM5592488</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Fed_SC 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501053</th>\n",
       "      <td>501053</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Feb 06 2017'</td>\n",
       "      <td>b'BCR-ABL transformed in vitro cell culture'</td>\n",
       "      <td>b'GSE39160'</td>\n",
       "      <td>b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...</td>\n",
       "      <td>145311674</td>\n",
       "      <td>87715899</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>polyA-enriched RNA was purified with the Micro...</td>\n",
       "      <td>strain: C57BL/6\\tgenotype/variation: Ikaros_Zn...</td>\n",
       "      <td>Raw data (fastq) was uploaded to the Galaxy we...</td>\n",
       "      <td>GSM957182</td>\n",
       "      <td>SRA</td>\n",
       "      <td>BCR-ABL+_Ikaros_ZnF4-/-_d14_repl1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>492764 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0 taxid_ch1                    status  \\\n",
       "0                0  b'10090'  b'Public on Apr 07 2015'   \n",
       "1                1  b'10090'  b'Public on Jun 10 2016'   \n",
       "2                2  b'10090'  b'Public on Jun 10 2016'   \n",
       "3                3  b'10090'  b'Public on Jan 01 2016'   \n",
       "4                4  b'10090'  b'Public on Sep 10 2012'   \n",
       "...            ...       ...                       ...   \n",
       "501049      501049  b'10090'  b'Public on Sep 23 2021'   \n",
       "501050      501050  b'10090'  b'Public on Sep 23 2021'   \n",
       "501051      501051  b'10090'  b'Public on Sep 23 2021'   \n",
       "501052      501052  b'10090'  b'Public on Sep 23 2021'   \n",
       "501053      501053  b'10090'  b'Public on Feb 06 2017'   \n",
       "\n",
       "                                     source_name_ch1  \\\n",
       "0                                   b'kidney cortex'   \n",
       "1                                 b'corpus callosum'   \n",
       "2                                     b'dorsal horn'   \n",
       "3                                     b'Lung tissue'   \n",
       "4                                          b'Cortex'   \n",
       "...                                              ...   \n",
       "501049          b'transfected with shGAA lentivirus'   \n",
       "501050               b'transfected with shscrambled'   \n",
       "501051               b'transfected with shscrambled'   \n",
       "501052               b'transfected with shscrambled'   \n",
       "501053  b'BCR-ABL transformed in vitro cell culture'   \n",
       "\n",
       "                              series_id  \\\n",
       "0                           b'GSE54390'   \n",
       "1                           b'GSE75330'   \n",
       "2                           b'GSE75330'   \n",
       "3                           b'GSE64027'   \n",
       "4       b'GSE36025\\tGSE49417\\tGSE49847'   \n",
       "...                                 ...   \n",
       "501049                     b'GSE184555'   \n",
       "501050                     b'GSE184555'   \n",
       "501051                     b'GSE184555'   \n",
       "501052                     b'GSE184555'   \n",
       "501053                      b'GSE39160'   \n",
       "\n",
       "                                                 relation  readstotal  \\\n",
       "0       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    39116064   \n",
       "1       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1136541   \n",
       "2       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      514285   \n",
       "3       b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    26804780   \n",
       "4       b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...   265978941   \n",
       "...                                                   ...         ...   \n",
       "501049  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    13275033   \n",
       "501050  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    12754218   \n",
       "501051  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    12243444   \n",
       "501052  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    12163822   \n",
       "501053  b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...   145311674   \n",
       "\n",
       "        readsaligned platform_id     organism_ch1  ... library_strategy.1  \\\n",
       "0           32884497    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "1             308433    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "2             148668    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "3           23355974    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "4          173619445    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "...              ...         ...              ...  ...                ...   \n",
       "501049      12476567    GPL21626  b'Mus musculus'  ...            RNA-Seq   \n",
       "501050      11967090    GPL21626  b'Mus musculus'  ...            RNA-Seq   \n",
       "501051      11535137    GPL21626  b'Mus musculus'  ...            RNA-Seq   \n",
       "501052      11405087    GPL21626  b'Mus musculus'  ...            RNA-Seq   \n",
       "501053      87715899    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "\n",
       "       library_selection.1 library_source.1   instrument_model.1  \\\n",
       "0                     cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "1                     cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "2                     cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "3                     cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "4                     cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "...                    ...              ...                  ...   \n",
       "501049                cDNA   transcriptomic          NextSeq 550   \n",
       "501050                cDNA   transcriptomic          NextSeq 550   \n",
       "501051                cDNA   transcriptomic          NextSeq 550   \n",
       "501052                cDNA   transcriptomic          NextSeq 550   \n",
       "501053                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "\n",
       "                                   extract_protocol_ch1.1  \\\n",
       "0       Thin outer slice of kidney cortex was immediat...   \n",
       "1       Selected CNS regions were dissected in 300µm t...   \n",
       "2       Selected CNS regions were dissected in 300µm t...   \n",
       "3       Total RNA was extracted from pooled dissected ...   \n",
       "4       http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...   \n",
       "...                                                   ...   \n",
       "501049  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501050  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501051  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501052  RNA was extracted using a QIAGEN Rneasy miniki...   \n",
       "501053  polyA-enriched RNA was purified with the Micro...   \n",
       "\n",
       "                                    characteristics_ch1.1  \\\n",
       "0       tissue: kidney cortex\\tstrain: C57BL/6\\tgender...   \n",
       "1       strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...   \n",
       "2       strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...   \n",
       "3       strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...   \n",
       "4       lab: CSHL-m\\tlab description: Gingeras - Cold ...   \n",
       "...                                                   ...   \n",
       "501049  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501050  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501051  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501052  cell line: C2C12\\tdifferentiation protocol: Di...   \n",
       "501053  strain: C57BL/6\\tgenotype/variation: Ikaros_Zn...   \n",
       "\n",
       "                                        data_processing.1 geo_accession.1  \\\n",
       "0       Reads were aligned to the genome using the Bur...      GSM1314260   \n",
       "1       Read processing was performed as described (Is...      GSM2164318   \n",
       "2       Read processing was performed as described (Is...      GSM2161048   \n",
       "3       Illumina Casava1.8 software used for basecalli...      GSM1563082   \n",
       "4       http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...      GSM1000563   \n",
       "...                                                   ...             ...   \n",
       "501049  Raw sequence that passed CASAV 1.8P/F filter w...      GSM5592485   \n",
       "501050  Raw sequence that passed CASAV 1.8P/F filter w...      GSM5592486   \n",
       "501051  Raw sequence that passed CASAV 1.8P/F filter w...      GSM5592487   \n",
       "501052  Raw sequence that passed CASAV 1.8P/F filter w...      GSM5592488   \n",
       "501053  Raw data (fastq) was uploaded to the Galaxy we...       GSM957182   \n",
       "\n",
       "       type.1                                            title.1  \n",
       "0         SRA                               TRACK 43-75 kidney 1  \n",
       "1         SRA                                 C1-1772117-029-A02  \n",
       "2         SRA                                 C1-1772072-242-F03  \n",
       "3         SRA                                 20NNK_LPS_Sil_I3C9  \n",
       "4         SRA  CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...  \n",
       "...       ...                                                ...  \n",
       "501049    SRA                                          Fed_GAA 3  \n",
       "501050    SRA                                           Fed_SC 1  \n",
       "501051    SRA                                           Fed_SC 2  \n",
       "501052    SRA                                           Fed_SC 3  \n",
       "501053    SRA                  BCR-ABL+_Ikaros_ZnF4-/-_d14_repl1  \n",
       "\n",
       "[492764 rows x 41 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del_single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a261900f-1af7-42c9-9f59-41ae13c46487",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "501054\n",
      "  taxid_ch1                    status     source_name_ch1  \\\n",
      "0  b'10090'  b'Public on Apr 07 2015'    b'kidney cortex'   \n",
      "1  b'10090'  b'Public on Jun 10 2016'  b'corpus callosum'   \n",
      "2  b'10090'  b'Public on Jun 10 2016'      b'dorsal horn'   \n",
      "3  b'10090'  b'Public on Jan 01 2016'      b'Lung tissue'   \n",
      "4  b'10090'  b'Public on Sep 10 2012'           b'Cortex'   \n",
      "\n",
      "                         series_id  \\\n",
      "0                      b'GSE54390'   \n",
      "1                      b'GSE75330'   \n",
      "2                      b'GSE75330'   \n",
      "3                      b'GSE64027'   \n",
      "4  b'GSE36025\\tGSE49417\\tGSE49847'   \n",
      "\n",
      "                                            relation  readstotal  \\\n",
      "0  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    39116064   \n",
      "1  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1136541   \n",
      "2  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      514285   \n",
      "3  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    26804780   \n",
      "4  b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...   265978941   \n",
      "\n",
      "   readsaligned platform_id     organism_ch1  molecule_ch1  ...  \\\n",
      "0      32884497    GPL13112  b'Mus musculus'  b'total RNA'  ...   \n",
      "1        308433    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
      "2        148668    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
      "3      23355974    GPL13112  b'Mus musculus'  b'total RNA'  ...   \n",
      "4     173619445    GPL13112  b'Mus musculus'  b'polyA RNA'  ...   \n",
      "\n",
      "  library_strategy library_selection  library_source     instrument_model  \\\n",
      "0          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "1          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "2          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "3          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "4          RNA-Seq              cDNA  transcriptomic  Illumina HiSeq 2000   \n",
      "\n",
      "                                extract_protocol_ch1  \\\n",
      "0  Thin outer slice of kidney cortex was immediat...   \n",
      "1  Selected CNS regions were dissected in 300µm t...   \n",
      "2  Selected CNS regions were dissected in 300µm t...   \n",
      "3  Total RNA was extracted from pooled dissected ...   \n",
      "4  http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...   \n",
      "\n",
      "                                 characteristics_ch1  \\\n",
      "0  tissue: kidney cortex\\tstrain: C57BL/6\\tgender...   \n",
      "1  strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...   \n",
      "2  strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...   \n",
      "3  strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...   \n",
      "4  lab: CSHL-m\\tlab description: Gingeras - Cold ...   \n",
      "\n",
      "                                     data_processing geo_accession type  \\\n",
      "0  Reads were aligned to the genome using the Bur...    GSM1314260  SRA   \n",
      "1  Read processing was performed as described (Is...    GSM2164318  SRA   \n",
      "2  Read processing was performed as described (Is...    GSM2161048  SRA   \n",
      "3  Illumina Casava1.8 software used for basecalli...    GSM1563082  SRA   \n",
      "4  http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...    GSM1000563  SRA   \n",
      "\n",
      "                                               title  \n",
      "0                               TRACK 43-75 kidney 1  \n",
      "1                                 C1-1772117-029-A02  \n",
      "2                                 C1-1772072-242-F03  \n",
      "3                                 20NNK_LPS_Sil_I3C9  \n",
      "4  CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...  \n",
      "\n",
      "[5 rows x 40 columns]\n"
     ]
    }
   ],
   "source": [
    "print(len(df))\n",
    "print(table.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ae37cf-fb4a-4b6d-8ad3-7280d54a7e9d",
   "metadata": {},
   "source": [
    "Create a table with only immune cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "68d8e482-43d5-49a4-b27d-5e3924133768",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_1162/1694858263.py:1: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df=pd.read_csv('del_single.csv')\n"
     ]
    }
   ],
   "source": [
    "df=pd.read_csv('del_single.csv')\n",
    "immune_cell= df[df['source_name_ch1'].str.contains('Innate lymphoid cell|CD8|CD4|macrophage|microglia|Th17|T cell|B cell', case=False)] # можно усложнить до регулярного выражения !\n",
    "\n",
    "immune_cell.to_csv('immune_cell.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4e962987-0925-4a59-adc6-5f532d965d89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>status</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>series_id</th>\n",
       "      <th>relation</th>\n",
       "      <th>readstotal</th>\n",
       "      <th>readsaligned</th>\n",
       "      <th>platform_id</th>\n",
       "      <th>...</th>\n",
       "      <th>library_strategy.1</th>\n",
       "      <th>library_selection.1</th>\n",
       "      <th>library_source.1</th>\n",
       "      <th>instrument_model.1</th>\n",
       "      <th>extract_protocol_ch1.1</th>\n",
       "      <th>characteristics_ch1.1</th>\n",
       "      <th>data_processing.1</th>\n",
       "      <th>geo_accession.1</th>\n",
       "      <th>type.1</th>\n",
       "      <th>title.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jan 30 2015'</td>\n",
       "      <td>b'Primary bone marrow-derived macrophages (BMDM)'</td>\n",
       "      <td>b'GSE56122\\tGSE56123'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>127650415</td>\n",
       "      <td>103088157</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was extracted from 5x106 cells using...</td>\n",
       "      <td>strain: BXH2/TyJ\\tcell type: BMDM cells (7th d...</td>\n",
       "      <td>Reads were quality filtered according to the I...</td>\n",
       "      <td>GSM1356226</td>\n",
       "      <td>SRA</td>\n",
       "      <td>BXH2_LPS_UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Mar 24 2015'</td>\n",
       "      <td>b'Peritoneal exudate cells (adherent cells)'</td>\n",
       "      <td>b'GSE55385'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>23911641</td>\n",
       "      <td>21984761</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>TRIzol\\tRNA libraries were prepared for sequen...</td>\n",
       "      <td>treatment: None\\tstrain: C57Bl6cJ</td>\n",
       "      <td>rsem v1.2.1/bowtie v0.12.9 was used to align t...</td>\n",
       "      <td>GSM1335239</td>\n",
       "      <td>SRA</td>\n",
       "      <td>PEC macrophage -IL10 -LPS replicate 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Mar 07 2016'</td>\n",
       "      <td>b'E12_PS34CD45-'</td>\n",
       "      <td>b'GSE54574'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>3258324</td>\n",
       "      <td>1377372</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Cells were sorted by FACS. Single cells isolat...</td>\n",
       "      <td>strain: c57bl6\\tage: embryonic day E12.5\\tgeno...</td>\n",
       "      <td>Illumina Casava 1.8.2 software used for baseca...</td>\n",
       "      <td>GSM1890434</td>\n",
       "      <td>SRA</td>\n",
       "      <td>E12_PS34CD45-_34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jul 15 2016'</td>\n",
       "      <td>b'Pro B cells'</td>\n",
       "      <td>b'GSE72018\\tGSE72019'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>50911147</td>\n",
       "      <td>45076587</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Female mouse C57BL/6 B cell populations were s...</td>\n",
       "      <td>strain: C57BL/6JNimr\\ttissue: Bone marrow\\tage...</td>\n",
       "      <td>fastq files were created with CASAVA 1.8.1\\tSe...</td>\n",
       "      <td>GSM1849922</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Bcell-pro-R5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 20 2016'</td>\n",
       "      <td>b'Follicular B cells'</td>\n",
       "      <td>b'GSE73532\\tGSE73534'</td>\n",
       "      <td>b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...</td>\n",
       "      <td>12791738</td>\n",
       "      <td>4090516</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>For bone marrow subset purification we used B6...</td>\n",
       "      <td>strain: C57Bl/6\\tage: 6-14 weeks\\tgender: pool...</td>\n",
       "      <td>Raw data were aligned to the mouse reference g...</td>\n",
       "      <td>GSM1897413</td>\n",
       "      <td>SRA</td>\n",
       "      <td>FOB RNA-seq_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492669</th>\n",
       "      <td>500959</td>\n",
       "      <td>500959</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'fibroblast cells'</td>\n",
       "      <td>b'GSE183704'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>11131614</td>\n",
       "      <td>8544928</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Qiazol (Qiagen) was added directly to PBS rins...</td>\n",
       "      <td>cell line: NIH3T3\\tsirna: CFIm59 si2</td>\n",
       "      <td>Raw reads were preprocessed to remove any adap...</td>\n",
       "      <td>GSM5568100</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CFIm59 si2 passage 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492670</th>\n",
       "      <td>500960</td>\n",
       "      <td>500960</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'fibroblast cells'</td>\n",
       "      <td>b'GSE183704'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>11655915</td>\n",
       "      <td>8742565</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Qiazol (Qiagen) was added directly to PBS rins...</td>\n",
       "      <td>cell line: NIH3T3\\tsirna: CFIm68 si1</td>\n",
       "      <td>Raw reads were preprocessed to remove any adap...</td>\n",
       "      <td>GSM5568101</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CFIm68 si1 passage 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492671</th>\n",
       "      <td>500961</td>\n",
       "      <td>500961</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'fibroblast cells'</td>\n",
       "      <td>b'GSE183704'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>9687597</td>\n",
       "      <td>7220444</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Qiazol (Qiagen) was added directly to PBS rins...</td>\n",
       "      <td>cell line: NIH3T3\\tsirna: CFIm68 si1</td>\n",
       "      <td>Raw reads were preprocessed to remove any adap...</td>\n",
       "      <td>GSM5568102</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CFIm68 si1 passage 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492672</th>\n",
       "      <td>500962</td>\n",
       "      <td>500962</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'fibroblast cells'</td>\n",
       "      <td>b'GSE183704'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>11954279</td>\n",
       "      <td>8987427</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Qiazol (Qiagen) was added directly to PBS rins...</td>\n",
       "      <td>cell line: NIH3T3\\tsirna: CFIm68 si2</td>\n",
       "      <td>Raw reads were preprocessed to remove any adap...</td>\n",
       "      <td>GSM5568104</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CFIm68 si2 passage 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>492673</th>\n",
       "      <td>500963</td>\n",
       "      <td>500963</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'fibroblast cells'</td>\n",
       "      <td>b'GSE183704'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>32182276</td>\n",
       "      <td>23802951</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Qiazol (Qiagen) was added directly to PBS rins...</td>\n",
       "      <td>cell line: NIH3T3\\tsirna: CFIm68 si2</td>\n",
       "      <td>Raw reads were preprocessed to remove any adap...</td>\n",
       "      <td>GSM5568105</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CFIm68 si2 passage 4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29791 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0.1  Unnamed: 0 taxid_ch1                    status  \\\n",
       "19                19          19  b'10090'  b'Public on Jan 30 2015'   \n",
       "24                24          24  b'10090'  b'Public on Mar 24 2015'   \n",
       "27                27          27  b'10090'  b'Public on Mar 07 2016'   \n",
       "28                28          28  b'10090'  b'Public on Jul 15 2016'   \n",
       "38                39          39  b'10090'  b'Public on Jun 20 2016'   \n",
       "...              ...         ...       ...                       ...   \n",
       "492669        500959      500959  b'10090'  b'Public on Sep 10 2021'   \n",
       "492670        500960      500960  b'10090'  b'Public on Sep 10 2021'   \n",
       "492671        500961      500961  b'10090'  b'Public on Sep 10 2021'   \n",
       "492672        500962      500962  b'10090'  b'Public on Sep 10 2021'   \n",
       "492673        500963      500963  b'10090'  b'Public on Sep 10 2021'   \n",
       "\n",
       "                                          source_name_ch1  \\\n",
       "19      b'Primary bone marrow-derived macrophages (BMDM)'   \n",
       "24           b'Peritoneal exudate cells (adherent cells)'   \n",
       "27                                       b'E12_PS34CD45-'   \n",
       "28                                         b'Pro B cells'   \n",
       "38                                  b'Follicular B cells'   \n",
       "...                                                   ...   \n",
       "492669                                b'fibroblast cells'   \n",
       "492670                                b'fibroblast cells'   \n",
       "492671                                b'fibroblast cells'   \n",
       "492672                                b'fibroblast cells'   \n",
       "492673                                b'fibroblast cells'   \n",
       "\n",
       "                    series_id  \\\n",
       "19      b'GSE56122\\tGSE56123'   \n",
       "24                b'GSE55385'   \n",
       "27                b'GSE54574'   \n",
       "28      b'GSE72018\\tGSE72019'   \n",
       "38      b'GSE73532\\tGSE73534'   \n",
       "...                       ...   \n",
       "492669           b'GSE183704'   \n",
       "492670           b'GSE183704'   \n",
       "492671           b'GSE183704'   \n",
       "492672           b'GSE183704'   \n",
       "492673           b'GSE183704'   \n",
       "\n",
       "                                                 relation  readstotal  \\\n",
       "19      b'BioSample: https://www.ncbi.nlm.nih.gov/bios...   127650415   \n",
       "24      b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    23911641   \n",
       "27      b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     3258324   \n",
       "28      b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    50911147   \n",
       "38      b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...    12791738   \n",
       "...                                                   ...         ...   \n",
       "492669  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    11131614   \n",
       "492670  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    11655915   \n",
       "492671  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     9687597   \n",
       "492672  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    11954279   \n",
       "492673  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    32182276   \n",
       "\n",
       "        readsaligned platform_id  ... library_strategy.1 library_selection.1  \\\n",
       "19         103088157    GPL13112  ...            RNA-Seq                cDNA   \n",
       "24          21984761    GPL13112  ...            RNA-Seq                cDNA   \n",
       "27           1377372    GPL13112  ...            RNA-Seq                cDNA   \n",
       "28          45076587    GPL13112  ...            RNA-Seq                cDNA   \n",
       "38           4090516    GPL13112  ...            RNA-Seq                cDNA   \n",
       "...              ...         ...  ...                ...                 ...   \n",
       "492669       8544928    GPL19057  ...            RNA-Seq                cDNA   \n",
       "492670       8742565    GPL19057  ...            RNA-Seq                cDNA   \n",
       "492671       7220444    GPL19057  ...            RNA-Seq                cDNA   \n",
       "492672       8987427    GPL19057  ...            RNA-Seq                cDNA   \n",
       "492673      23802951    GPL19057  ...            RNA-Seq                cDNA   \n",
       "\n",
       "       library_source.1    instrument_model.1  \\\n",
       "19       transcriptomic   Illumina HiSeq 2000   \n",
       "24       transcriptomic   Illumina HiSeq 2000   \n",
       "27       transcriptomic   Illumina HiSeq 2000   \n",
       "28       transcriptomic   Illumina HiSeq 2000   \n",
       "38       transcriptomic   Illumina HiSeq 2000   \n",
       "...                 ...                   ...   \n",
       "492669   transcriptomic  Illumina NextSeq 500   \n",
       "492670   transcriptomic  Illumina NextSeq 500   \n",
       "492671   transcriptomic  Illumina NextSeq 500   \n",
       "492672   transcriptomic  Illumina NextSeq 500   \n",
       "492673   transcriptomic  Illumina NextSeq 500   \n",
       "\n",
       "                                   extract_protocol_ch1.1  \\\n",
       "19      Total RNA was extracted from 5x106 cells using...   \n",
       "24      TRIzol\\tRNA libraries were prepared for sequen...   \n",
       "27      Cells were sorted by FACS. Single cells isolat...   \n",
       "28      Female mouse C57BL/6 B cell populations were s...   \n",
       "38      For bone marrow subset purification we used B6...   \n",
       "...                                                   ...   \n",
       "492669  Qiazol (Qiagen) was added directly to PBS rins...   \n",
       "492670  Qiazol (Qiagen) was added directly to PBS rins...   \n",
       "492671  Qiazol (Qiagen) was added directly to PBS rins...   \n",
       "492672  Qiazol (Qiagen) was added directly to PBS rins...   \n",
       "492673  Qiazol (Qiagen) was added directly to PBS rins...   \n",
       "\n",
       "                                    characteristics_ch1.1  \\\n",
       "19      strain: BXH2/TyJ\\tcell type: BMDM cells (7th d...   \n",
       "24                      treatment: None\\tstrain: C57Bl6cJ   \n",
       "27      strain: c57bl6\\tage: embryonic day E12.5\\tgeno...   \n",
       "28      strain: C57BL/6JNimr\\ttissue: Bone marrow\\tage...   \n",
       "38      strain: C57Bl/6\\tage: 6-14 weeks\\tgender: pool...   \n",
       "...                                                   ...   \n",
       "492669               cell line: NIH3T3\\tsirna: CFIm59 si2   \n",
       "492670               cell line: NIH3T3\\tsirna: CFIm68 si1   \n",
       "492671               cell line: NIH3T3\\tsirna: CFIm68 si1   \n",
       "492672               cell line: NIH3T3\\tsirna: CFIm68 si2   \n",
       "492673               cell line: NIH3T3\\tsirna: CFIm68 si2   \n",
       "\n",
       "                                        data_processing.1 geo_accession.1  \\\n",
       "19      Reads were quality filtered according to the I...      GSM1356226   \n",
       "24      rsem v1.2.1/bowtie v0.12.9 was used to align t...      GSM1335239   \n",
       "27      Illumina Casava 1.8.2 software used for baseca...      GSM1890434   \n",
       "28      fastq files were created with CASAVA 1.8.1\\tSe...      GSM1849922   \n",
       "38      Raw data were aligned to the mouse reference g...      GSM1897413   \n",
       "...                                                   ...             ...   \n",
       "492669  Raw reads were preprocessed to remove any adap...      GSM5568100   \n",
       "492670  Raw reads were preprocessed to remove any adap...      GSM5568101   \n",
       "492671  Raw reads were preprocessed to remove any adap...      GSM5568102   \n",
       "492672  Raw reads were preprocessed to remove any adap...      GSM5568104   \n",
       "492673  Raw reads were preprocessed to remove any adap...      GSM5568105   \n",
       "\n",
       "       type.1                                title.1  \n",
       "19        SRA                            BXH2_LPS_UT  \n",
       "24        SRA  PEC macrophage -IL10 -LPS replicate 2  \n",
       "27        SRA                       E12_PS34CD45-_34  \n",
       "28        SRA                           Bcell-pro-R5  \n",
       "38        SRA                         FOB RNA-seq_r1  \n",
       "...       ...                                    ...  \n",
       "492669    SRA                   CFIm59 si2 passage 8  \n",
       "492670    SRA                   CFIm68 si1 passage 3  \n",
       "492671    SRA                   CFIm68 si1 passage 4  \n",
       "492672    SRA                   CFIm68 si2 passage 3  \n",
       "492673    SRA                   CFIm68 si2 passage 4  \n",
       "\n",
       "[29791 rows x 42 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "immune_cell"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "128d2b54-491b-4e85-8682-484f35b91aa1",
   "metadata": {},
   "source": [
    "Create a table with uniqe values by cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a136bb32-eb33-46ec-b637-3315bd7ea415",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_1162/1197267492.py:1: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('del_single.csv')\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('del_single.csv')\n",
    "cell_count=df.groupby('source_name_ch1')['platform_id'].value_counts() #count on two column basis\n",
    "#cell_count=df['source_name_ch1'].value_counts() #create a table with count of uniq cell types\n",
    "cell_count.to_csv('cell_count.csv') # put table into file .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ec03a630-dbf0-4544-99cf-419a8260036b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "source_name_ch1                                                         platform_id\n",
       "b\"2W-Salmonella-infected Peyer's patches, mesenteric lymph node cells\"  GPL19057         6\n",
       "b\"3'-Seq in OPMD mouse model #1; age of 12 months\"                      GPL13112         1\n",
       "b\"3'-Seq in OPMD mouse model #1; age of 20 weeks\"                       GPL13112         1\n",
       "b\"3'-Seq in OPMD mouse model #2; age of 12 months\"                      GPL13112         1\n",
       "b\"3'-Seq in OPMD mouse model #2; age of 20 weeks\"                       GPL13112         1\n",
       "                                                                                      ... \n",
       "b'young_naive_Mono_Saline'                                              GPL19057         5\n",
       "b'zona incerta'                                                         GPL13112       187\n",
       "b'zygote from Cxxc1fl/fl;Zp3-Cre female and wildtype male'              GPL21273         2\n",
       "b'zygote from wildtype female and wildtype male'                        GPL21273         3\n",
       "b'zygote'                                                               GPL17021        15\n",
       "Name: platform_id, Length: 38323, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36031c8e-1280-46fd-8987-ec6201e522ea",
   "metadata": {},
   "source": [
    "Create a table with uniqe values by platform id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "745342ec-b7fd-4640-91fb-8a4b36489564",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_1162/137564493.py:1: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('del_single.csv')\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('del_single.csv')\n",
    "platform_count=df.groupby('platform_id')['source_name_ch1'].value_counts() #count on two column basis\n",
    "#cell_count=df['source_name_ch1'].value_counts() #create a table with count of uniq cell types\n",
    "platform_count.to_csv('platform_count.csv') # put table into file .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "fc818770-ac02-484d-a3e9-797c8c299c7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "platform_id  source_name_ch1    \n",
       "GPL13112     b'Brain'               7605\n",
       "             b'hypothalamus'        3104\n",
       "             b'Liver'               2751\n",
       "             b'Embryo'              1846\n",
       "             b'ventral midbrain'    1845\n",
       "                                    ... \n",
       "GPL24247     b'wt-1'                   1\n",
       "             b'wt-2'                   1\n",
       "             b'wt-3'                   1\n",
       "             b'wt-4'                   1\n",
       "missing      b'missing'             2036\n",
       "Name: source_name_ch1, Length: 38323, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "platform_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbfa3a85-84bd-49b5-94e0-b692faac88bb",
   "metadata": {},
   "source": [
    "Create a table with uniqe values by cell type - only immune cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3673be56-e88e-48ff-9fc0-ac23be219ffb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('immune_cell.csv')\n",
    "cell_count=df.groupby('source_name_ch1')['platform_id'].value_counts() #count on two column basis\n",
    "#cell_count=df['source_name_ch1'].value_counts() #create a table with count of uniq cell types\n",
    "cell_count.to_csv('cell_count2.csv') # put table into file .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2745e560-1ed3-489b-a36d-a4d8a327a21c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "source_name_ch1                                                                          platform_id\n",
       "b\"Alveolar macrophage  from flu infected mouse's lung\"                                   GPL18480        3\n",
       "b\"Bone marrow derived macrophage- stimulated 30'/DMSO treated\"                           GPL13112       11\n",
       "b\"CD19+IgD-B220+IgM+ 3'Ek-/- immature B cells\"                                           GPL13112        2\n",
       "b\"CD19+IgD-B220+IgM-CD25+ckit- 3'Ek-/- pre-B cells\"                                      GPL13112        2\n",
       "b\"CD4+CD8+ double positive 3'Ek-/- T cells\"                                              GPL13112        2\n",
       "                                                                                                        ..\n",
       "b'wild-type C57BL/6 mouse peritoneal macrophages'                                        GPL15103        6\n",
       "b'wildtype ProB cells (B220+CD43+IgM-)'                                                  GPL13112        2\n",
       "b'wt proB cells, Replicate1'                                                             GPL24247        1\n",
       "b'wt proB cells, Replicate2'                                                             GPL24247        1\n",
       "b'young (4 month) alveolar macrophages adoptively transferred into old mice (18 month)'  GPL19057        3\n",
       "Name: platform_id, Length: 3141, dtype: int64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bd32bd0-1342-4694-9971-5ab4d12ed48d",
   "metadata": {},
   "source": [
    "Create a table with uniqe values by platform id - only immune cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "bb2bb2de-68a5-42a3-8be1-206bd47721a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('immune_cell.csv')\n",
    "platform_count=df.groupby('platform_id')['source_name_ch1'].value_counts() #count on two column basis\n",
    "#cell_count=df['source_name_ch1'].value_counts() #create a table with count of uniq cell types\n",
    "platform_count.to_csv('platform_count2.csv') # put table into file .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5bcc2df1-deb0-4f30-b1f5-38f1e1251b35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "platform_id  source_name_ch1                                  \n",
       "GPL13112     b'microglia'                                         1023\n",
       "             b'bone marrow-derived macrophages'                    477\n",
       "             b'Bone Marrow Derived Macrophages'                     90\n",
       "             b'CD4+ T cells'                                        80\n",
       "             b'Primary bone marrow-derived macrophages (BMDM)'      77\n",
       "                                                                  ... \n",
       "GPL24247     b'infected macrophages'                                 1\n",
       "             b'intestinal CD45+ cells'                               1\n",
       "             b'peritoneal macrophage'                                1\n",
       "             b'wt proB cells, Replicate1'                            1\n",
       "             b'wt proB cells, Replicate2'                            1\n",
       "Name: source_name_ch1, Length: 3141, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "platform_count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f31ac6e6-1e8c-49ef-bf2f-3e3e66a04b19",
   "metadata": {},
   "source": [
    "Let's change the last table. Remove very small values, type of cells into rows, platform id into columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "8c01bcbd-5ab0-46c1-acc6-31ed6efe7294",
   "metadata": {},
   "outputs": [],
   "source": [
    "df= pd.read_csv('platform_count2.csv')\n",
    "df.columns=['platform_id', 'source_name_ch1', 'counter'] # rename the columns heads\n",
    "df=df[df.counter>10] # delete the values less than 10\n",
    "table = pd.pivot_table(df, values='counter', index=['source_name_ch1'], columns=['platform_id'], aggfunc=np.sum) #create a table with platfom_id in columns\n",
    "table.to_csv('platform_count3.csv') # put table into file .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "829d4b23-95eb-442f-9d5c-9ef4a8da13a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>platform_id</th>\n",
       "      <th>GPL13112</th>\n",
       "      <th>GPL15103</th>\n",
       "      <th>GPL17021</th>\n",
       "      <th>GPL18480</th>\n",
       "      <th>GPL19057</th>\n",
       "      <th>GPL20795</th>\n",
       "      <th>GPL21103</th>\n",
       "      <th>GPL21273</th>\n",
       "      <th>GPL21493</th>\n",
       "      <th>GPL21626</th>\n",
       "      <th>GPL23479</th>\n",
       "      <th>GPL24247</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b\"Bone marrow derived macrophage- stimulated 30'/DMSO treated\"</th>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'AV node macrophage'</th>\n",
       "      <td>76.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'Act B cell'</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'Activated CD4+ T cells'</th>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'Activated T cells'</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'thioglycolate-induced peritoneal macrophages'</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>44.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'thioglycollate elicited macrophage cells'</th>\n",
       "      <td>18.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'tumor infiltrating macrophage'</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'tumor macrophages'</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b'tumor-infiltrating CD8+ T lymphocyte'</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>366 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "platform_id                                         GPL13112  GPL15103  \\\n",
       "source_name_ch1                                                          \n",
       "b\"Bone marrow derived macrophage- stimulated 30...      11.0       NaN   \n",
       "b'AV node macrophage'                                   76.0       NaN   \n",
       "b'Act B cell'                                            NaN       NaN   \n",
       "b'Activated CD4+ T cells'                               13.0       NaN   \n",
       "b'Activated T cells'                                     NaN       NaN   \n",
       "...                                                      ...       ...   \n",
       "b'thioglycolate-induced peritoneal macrophages'          NaN       NaN   \n",
       "b'thioglycollate elicited macrophage cells'             18.0       NaN   \n",
       "b'tumor infiltrating macrophage'                         NaN       NaN   \n",
       "b'tumor macrophages'                                     NaN       NaN   \n",
       "b'tumor-infiltrating CD8+ T lymphocyte'                  NaN       NaN   \n",
       "\n",
       "platform_id                                         GPL17021  GPL18480  \\\n",
       "source_name_ch1                                                          \n",
       "b\"Bone marrow derived macrophage- stimulated 30...       NaN       NaN   \n",
       "b'AV node macrophage'                                    NaN       NaN   \n",
       "b'Act B cell'                                           24.0       NaN   \n",
       "b'Activated CD4+ T cells'                                NaN       NaN   \n",
       "b'Activated T cells'                                     NaN       NaN   \n",
       "...                                                      ...       ...   \n",
       "b'thioglycolate-induced peritoneal macrophages'          NaN       NaN   \n",
       "b'thioglycollate elicited macrophage cells'              NaN       NaN   \n",
       "b'tumor infiltrating macrophage'                         NaN       NaN   \n",
       "b'tumor macrophages'                                     NaN       NaN   \n",
       "b'tumor-infiltrating CD8+ T lymphocyte'                  NaN       NaN   \n",
       "\n",
       "platform_id                                         GPL19057  GPL20795  \\\n",
       "source_name_ch1                                                          \n",
       "b\"Bone marrow derived macrophage- stimulated 30...       NaN       NaN   \n",
       "b'AV node macrophage'                                    NaN       NaN   \n",
       "b'Act B cell'                                            NaN       NaN   \n",
       "b'Activated CD4+ T cells'                                NaN       NaN   \n",
       "b'Activated T cells'                                     NaN       NaN   \n",
       "...                                                      ...       ...   \n",
       "b'thioglycolate-induced peritoneal macrophages'          NaN       NaN   \n",
       "b'thioglycollate elicited macrophage cells'              NaN       NaN   \n",
       "b'tumor infiltrating macrophage'                         NaN       NaN   \n",
       "b'tumor macrophages'                                     NaN       NaN   \n",
       "b'tumor-infiltrating CD8+ T lymphocyte'                 13.0       NaN   \n",
       "\n",
       "platform_id                                         GPL21103  GPL21273  \\\n",
       "source_name_ch1                                                          \n",
       "b\"Bone marrow derived macrophage- stimulated 30...       NaN       NaN   \n",
       "b'AV node macrophage'                                    NaN       NaN   \n",
       "b'Act B cell'                                            NaN       NaN   \n",
       "b'Activated CD4+ T cells'                                NaN       NaN   \n",
       "b'Activated T cells'                                     NaN       NaN   \n",
       "...                                                      ...       ...   \n",
       "b'thioglycolate-induced peritoneal macrophages'          NaN       NaN   \n",
       "b'thioglycollate elicited macrophage cells'              NaN       NaN   \n",
       "b'tumor infiltrating macrophage'                         NaN      17.0   \n",
       "b'tumor macrophages'                                    12.0       NaN   \n",
       "b'tumor-infiltrating CD8+ T lymphocyte'                  NaN       NaN   \n",
       "\n",
       "platform_id                                         GPL21493  GPL21626  \\\n",
       "source_name_ch1                                                          \n",
       "b\"Bone marrow derived macrophage- stimulated 30...       NaN       NaN   \n",
       "b'AV node macrophage'                                    NaN       NaN   \n",
       "b'Act B cell'                                            NaN       NaN   \n",
       "b'Activated CD4+ T cells'                                NaN       NaN   \n",
       "b'Activated T cells'                                     NaN       NaN   \n",
       "...                                                      ...       ...   \n",
       "b'thioglycolate-induced peritoneal macrophages'         44.0       NaN   \n",
       "b'thioglycollate elicited macrophage cells'              NaN       NaN   \n",
       "b'tumor infiltrating macrophage'                         NaN       NaN   \n",
       "b'tumor macrophages'                                     NaN       NaN   \n",
       "b'tumor-infiltrating CD8+ T lymphocyte'                  NaN       NaN   \n",
       "\n",
       "platform_id                                         GPL23479  GPL24247  \n",
       "source_name_ch1                                                         \n",
       "b\"Bone marrow derived macrophage- stimulated 30...       NaN       NaN  \n",
       "b'AV node macrophage'                                    NaN       NaN  \n",
       "b'Act B cell'                                            NaN       NaN  \n",
       "b'Activated CD4+ T cells'                                NaN       NaN  \n",
       "b'Activated T cells'                                    24.0       NaN  \n",
       "...                                                      ...       ...  \n",
       "b'thioglycolate-induced peritoneal macrophages'          NaN       NaN  \n",
       "b'thioglycollate elicited macrophage cells'              NaN       NaN  \n",
       "b'tumor infiltrating macrophage'                         NaN       NaN  \n",
       "b'tumor macrophages'                                     NaN       NaN  \n",
       "b'tumor-infiltrating CD8+ T lymphocyte'                  NaN       NaN  \n",
       "\n",
       "[366 rows x 12 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82e4aa4d-c0a3-4357-a136-91f266c02ded",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_1184/787893585.py:1: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('del_single.csv')\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('del_single.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b375ec5-a7a7-4bb8-8238-8d5f281dbc62",
   "metadata": {},
   "source": [
    "Create a table with needing columns - макрофаги\n",
    "air_quality = pd.concat([air_quality_pm25, air_quality_no2], axis=0) - сложение таблиц"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e2984b96-5e86-42a1-be8a-0fb153c31686",
   "metadata": {},
   "outputs": [],
   "source": [
    "wf = df[['geo_accession', 'platform_id', 'source_name_ch1']] #wf - working frame\n",
    "macrophage= wf[wf['source_name_ch1'].str.contains('macrophage|macrophages', case=False)]\n",
    "macrophage.insert(2, \"Cell_type\", 'macrophage')\n",
    "conditions = [\n",
    "    (macrophage['source_name_ch1'].str.contains('[B,m]one[-,\\s]marrow', case=False, regex=True)),\n",
    "    (macrophage['source_name_ch1'].str.contains('Peritoneal', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('Alveolar|lung', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('tumor|tumour', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('Splenic', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('hepatic', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('sciatic nerve', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('Phagocytic', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('Peripheral', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('muscle', case=False)),\n",
    "    (macrophage['source_name_ch1'].str.contains('1 day after TBI', case=False)),\n",
    "    (~macrophage['source_name_ch1'].str.contains('[B,m]one[-,\\s]marrow|Peritoneal|Alveolar|tumor|tumour|Splenic|lung|hepatic|sciatic nerve|Phagocytic|Peripheral|muscle|1 day after TBI' , case=False))\n",
    "]\n",
    "\n",
    "values = ['Bone Marrow Derived macrophages', 'Peritoneal macrophage', 'Alveolar macrophage', 'tumor macrophage','Splenic macrophages','hepatic macrophages', 'sciatic nerve macrophages','Phagocytic macrophages', 'Peripheral macrophages','muscle macrophages','1 day after TBI macrophages', 'other macrophage']\n",
    "macrophage.insert(3, \"Classes\", np.select(conditions, values))\n",
    "macrophage.to_csv('macrophage.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "a627869d-256c-48d0-9d01-9f4ecc878327",
   "metadata": {},
   "outputs": [],
   "source": [
    "wf = df[['geo_accession', 'platform_id', 'source_name_ch1']] #wf - working frame\n",
    "T_cells= wf[wf['source_name_ch1'].str.contains('/sT[-,\\s)]cell[\\s,s]|\\sCD8[+,\\s][\\s]|[\\s]CD4[+,\\s][\\s]', case=False)]\n",
    "T_cells.insert(2, \"Cell_type\", 'T_cells')\n",
    "conditions = [\n",
    "    (T_cells['source_name_ch1'].str.contains('[\\s]CD8[+,\\s][\\s]',case=False, regex=True)),\n",
    "    (T_cells['source_name_ch1'].str.contains('[\\s]CD4[+,\\s][\\s]',case=False, regex=True)),\n",
    "    (T_cells['source_name_ch1'].str.contains('YFP+', case=False, regex=True)),\n",
    "    (T_cells['source_name_ch1'].str.contains('V\\xce\\xb34', case=False)),\n",
    "    (T_cells['source_name_ch1'].str.contains('Tumor-infiltrating', case=False)),\n",
    "    (T_cells['source_name_ch1'].str.contains('Thymus|Thymic', case=False)),\n",
    "    (T_cells['source_name_ch1'].str.contains('Splenic|spleen', case=False)),\n",
    "    (~T_cells['source_name_ch1'].str.contains('[\\s]CD8[+,\\s]|[\\s]CD4[+,\\s][\\s]|YFP+|V\\xce\\xb34|Tumor-infiltrating|Thymus|Thymic|Splenic|spleen', case=False, regex=True))\n",
    "]\n",
    "values = ['CD8+ T cells', 'CD4+ T cells', 'YFP+ T cells','V\\xce\\xb34 T cell', 'Tumor-infiltrating T cells', 'Thymus T cells', 'Splenic T cells', 'other T cells']\n",
    "T_cells.insert(3, \"Classes\", np.select(conditions, values))\n",
    "T_cells.to_csv('T_cells.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "24c4aa7e-aec0-4ae1-90c2-c01218c1d9f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "wf = df[['geo_accession', 'platform_id', 'source_name_ch1']] #wf - working frame\n",
    "B_cells= wf[wf['source_name_ch1'].str.contains('[\\s]B[-, \\s]cell[s,\\s]', case=False, regex=True)]\n",
    "B_cells.insert(2, \"Cell_type\", 'B_cells')\n",
    "conditions = [\n",
    "    (B_cells['source_name_ch1'].str.contains('splenic|from spleen|spleenic',case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('Germinal center',case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('Marginal zone' , case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('Follicular|folicular' , case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('B220+' , case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('Primary' , case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('NP+ GC', case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('immature', case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('CH12 ', case=False, regex=True)),\n",
    "    (B_cells['source_name_ch1'].str.contains('CD19+', case=False, regex=True)),\n",
    "    (~B_cells['source_name_ch1'].str.contains('splenic|from spleen|spleenic|Germinal center|Marginal zone|Follicular|folicular|B220+|primary|NP+ GC|immature|CH12|CD19+ ' , case=False, regex=True))\n",
    "\n",
    "]\n",
    "values = ['splenic B cells', 'Germinal center B cells', 'Marginal zone B cells', 'Follicular B cells', 'B220+ B cells','Primary B cells','NP+ GC B cells','Immature B cells','CH12 line','CD19+ B cells', 'other B-cells']\n",
    "B_cells.insert(3, \"Classes\", np.select(conditions, values))\n",
    "B_cells.to_csv('B_cells.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da8e42e7-5799-4c72-a073-3470fb47b319",
   "metadata": {},
   "source": [
    "# Создадим большую таблицу со всеми типами и классами - great_table->great_table2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f34596b3-eb47-4736-bf3c-e89f855c402c",
   "metadata": {},
   "outputs": [],
   "source": [
    "B_cells=pd.read_csv('B_cells.csv')\n",
    "T_cells=pd.read_csv('T_cells.csv')\n",
    "macrophage=pd.read_csv('macrophage.csv')\n",
    "great_table = pd.concat([B_cells, T_cells, macrophage], axis=0)\n",
    "\n",
    "great_table.to_csv('great_table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "edee1fc8-b0bb-4de8-90b0-3667ea18ddeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>platform_id</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>Classes</th>\n",
       "      <th>source_name_ch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28</td>\n",
       "      <td>GSM1849922</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>other B-cells</td>\n",
       "      <td>b'Pro B cells'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>38</td>\n",
       "      <td>GSM1897413</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>Follicular B cells</td>\n",
       "      <td>b'Follicular B cells'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>334</td>\n",
       "      <td>GSM1849954</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>other B-cells</td>\n",
       "      <td>b'B1A B cells'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>341</td>\n",
       "      <td>GSM1229014</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>other B-cells</td>\n",
       "      <td>b'Dnmt-KO B-Cells'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>570</td>\n",
       "      <td>GSM1234739</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>B_cells</td>\n",
       "      <td>Primary B cells</td>\n",
       "      <td>b'Primary B-cells'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6358</th>\n",
       "      <td>492513</td>\n",
       "      <td>GSM5550909</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>other macrophage</td>\n",
       "      <td>b'Macrophages'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6359</th>\n",
       "      <td>492636</td>\n",
       "      <td>GSM5558220</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>other macrophage</td>\n",
       "      <td>b'macrophages_WT'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6360</th>\n",
       "      <td>492637</td>\n",
       "      <td>GSM5558221</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>other macrophage</td>\n",
       "      <td>b'macrophages_WT'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6361</th>\n",
       "      <td>492638</td>\n",
       "      <td>GSM5558224</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>other macrophage</td>\n",
       "      <td>b'macrophages_Dnmt3a KO'</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6362</th>\n",
       "      <td>492639</td>\n",
       "      <td>GSM5558226</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>other macrophage</td>\n",
       "      <td>b'macrophages_Dnmt3a KO'</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9410 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0 geo_accession platform_id   Cell_type             Classes  \\\n",
       "0             28    GSM1849922    GPL13112     B_cells       other B-cells   \n",
       "1             38    GSM1897413    GPL13112     B_cells  Follicular B cells   \n",
       "2            334    GSM1849954    GPL13112     B_cells       other B-cells   \n",
       "3            341    GSM1229014    GPL13112     B_cells       other B-cells   \n",
       "4            570    GSM1234739    GPL13112     B_cells     Primary B cells   \n",
       "...          ...           ...         ...         ...                 ...   \n",
       "6358      492513    GSM5550909    GPL19057  macrophage    other macrophage   \n",
       "6359      492636    GSM5558220    GPL21626  macrophage    other macrophage   \n",
       "6360      492637    GSM5558221    GPL21626  macrophage    other macrophage   \n",
       "6361      492638    GSM5558224    GPL21626  macrophage    other macrophage   \n",
       "6362      492639    GSM5558226    GPL21626  macrophage    other macrophage   \n",
       "\n",
       "               source_name_ch1  \n",
       "0               b'Pro B cells'  \n",
       "1        b'Follicular B cells'  \n",
       "2               b'B1A B cells'  \n",
       "3           b'Dnmt-KO B-Cells'  \n",
       "4           b'Primary B-cells'  \n",
       "...                        ...  \n",
       "6358            b'Macrophages'  \n",
       "6359         b'macrophages_WT'  \n",
       "6360         b'macrophages_WT'  \n",
       "6361  b'macrophages_Dnmt3a KO'  \n",
       "6362  b'macrophages_Dnmt3a KO'  \n",
       "\n",
       "[9410 rows x 6 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "great_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87e8c858-07e5-45f0-b1ce-b12ab9b961bf",
   "metadata": {},
   "source": [
    "get indicis of lines with geo accession\n",
    "indices[i]=indices[i-1] можно так "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5ef4c89b-ea6f-49d4-9b56-57bfdd223d00",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_2926/3301459707.py:2: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('table.csv')\n"
     ]
    }
   ],
   "source": [
    "great_table=pd.read_csv('great_table.csv')\n",
    "df = pd.read_csv('table.csv')\n",
    "res=df[df.geo_accession.isin(great_table.geo_accession)].reset_index(drop=False)\n",
    "indices=res['index']\n",
    "indices2=[]\n",
    "for i in indices:\n",
    "    f=i\n",
    "    indices2.append(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fc334f64-4252-4156-8e19-a00d4d053c35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9410\n",
      "9410\n"
     ]
    }
   ],
   "source": [
    "print(len(great_table))\n",
    "print(len(indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "5d693792-5f53-4710-9ee4-2546893d8b86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>status</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>series_id</th>\n",
       "      <th>relation</th>\n",
       "      <th>readstotal</th>\n",
       "      <th>readsaligned</th>\n",
       "      <th>platform_id</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>library_strategy.1</th>\n",
       "      <th>library_selection.1</th>\n",
       "      <th>library_source.1</th>\n",
       "      <th>instrument_model.1</th>\n",
       "      <th>extract_protocol_ch1.1</th>\n",
       "      <th>characteristics_ch1.1</th>\n",
       "      <th>data_processing.1</th>\n",
       "      <th>geo_accession.1</th>\n",
       "      <th>type.1</th>\n",
       "      <th>title.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Apr 07 2015'</td>\n",
       "      <td>b'kidney cortex'</td>\n",
       "      <td>b'GSE54390'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>39116064</td>\n",
       "      <td>32884497</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Thin outer slice of kidney cortex was immediat...</td>\n",
       "      <td>tissue: kidney cortex\\tstrain: C57BL/6\\tgender...</td>\n",
       "      <td>Reads were aligned to the genome using the Bur...</td>\n",
       "      <td>GSM1314260</td>\n",
       "      <td>SRA</td>\n",
       "      <td>TRACK 43-75 kidney 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'corpus callosum'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1136541</td>\n",
       "      <td>308433</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2164318</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772117-029-A02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'dorsal horn'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>514285</td>\n",
       "      <td>148668</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2161048</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772072-242-F03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jan 01 2016'</td>\n",
       "      <td>b'Lung tissue'</td>\n",
       "      <td>b'GSE64027'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>26804780</td>\n",
       "      <td>23355974</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was extracted from pooled dissected ...</td>\n",
       "      <td>strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...</td>\n",
       "      <td>Illumina Casava1.8 software used for basecalli...</td>\n",
       "      <td>GSM1563082</td>\n",
       "      <td>SRA</td>\n",
       "      <td>20NNK_LPS_Sil_I3C9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2012'</td>\n",
       "      <td>b'Cortex'</td>\n",
       "      <td>b'GSE36025\\tGSE49417\\tGSE49847'</td>\n",
       "      <td>b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...</td>\n",
       "      <td>265978941</td>\n",
       "      <td>173619445</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...</td>\n",
       "      <td>lab: CSHL-m\\tlab description: Gingeras - Cold ...</td>\n",
       "      <td>http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...</td>\n",
       "      <td>GSM1000563</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Oct 27 2016'</td>\n",
       "      <td>b'E12.5 Single Cell'</td>\n",
       "      <td>b'GSE87069'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>5607362</td>\n",
       "      <td>1929834</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Smart-seq2\\tNextera XT DNA library preparation...</td>\n",
       "      <td>strain: C57BL/6 Heterozygous Lmx1a-EGFP\\tdevel...</td>\n",
       "      <td>reads were aligned to the mm10 genome using de...</td>\n",
       "      <td>GSM2320580</td>\n",
       "      <td>SRA</td>\n",
       "      <td>F125_RNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'corpus callosum'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1065010</td>\n",
       "      <td>473113</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2164283</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772117-028-A05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Feb 20 2015'</td>\n",
       "      <td>b'sscortex'</td>\n",
       "      <td>b'GSE60361'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1211044</td>\n",
       "      <td>330539</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Mice were deeply anesthetized with ketamine/xy...</td>\n",
       "      <td>Sex: male\\tage: P23</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM1474588</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SingleCellNo338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Oct 06 2016'</td>\n",
       "      <td>b'ventral midbrain'</td>\n",
       "      <td>b'GSE76381'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>663629</td>\n",
       "      <td>133904</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Human embryo ventral midbrain dissection was p...</td>\n",
       "      <td>tissue: ventral midbrain\\tSex: pooled male and...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2273393</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1772075301_G03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'SN-VTA'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>456330</td>\n",
       "      <td>131980</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: CD1\\tage: p20\\tSex: pooled male and fe...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2162146</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772096-087-D08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Feb 20 2015'</td>\n",
       "      <td>b'ca1hippocampus'</td>\n",
       "      <td>b'GSE60361'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>2534097</td>\n",
       "      <td>584659</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Mice were deeply anesthetized with ketamine/xy...</td>\n",
       "      <td>Sex: male\\tage: P23</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM1475854</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SingleCellNo1599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 10 2016'</td>\n",
       "      <td>b'SN-VTA'</td>\n",
       "      <td>b'GSE75330'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1362059</td>\n",
       "      <td>570834</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Selected CNS regions were dissected in 300µm t...</td>\n",
       "      <td>strain: CD1\\tage: p21\\tSex: pooled male and fe...</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM2161497</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C1-1772089-200-B02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Feb 20 2015'</td>\n",
       "      <td>b'sscortex'</td>\n",
       "      <td>b'GSE60361'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>3861394</td>\n",
       "      <td>1464176</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Mice were deeply anesthetized with ketamine/xy...</td>\n",
       "      <td>Sex: male\\tage: P26</td>\n",
       "      <td>Read processing was performed as described (Is...</td>\n",
       "      <td>GSM1474327</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SingleCellNo77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Apr 10 2013'</td>\n",
       "      <td>b'Dicer -/- murine MSC cell line'</td>\n",
       "      <td>b'GSE44157\\tGSE44163'</td>\n",
       "      <td>b'Reanalyzed by: GSE80797\\tSRA: https://www.nc...</td>\n",
       "      <td>27584764</td>\n",
       "      <td>25018842</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>mRNA-sequencing was carried out with total RNA...</td>\n",
       "      <td>cell type: murine MSC cell line\\tgenotype: Dic...</td>\n",
       "      <td>Illumina pipeline version 1.3+ used for baseca...</td>\n",
       "      <td>GSM1079794</td>\n",
       "      <td>SRA</td>\n",
       "      <td>KO_siNr6a1_R0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Apr 16 2014'</td>\n",
       "      <td>b'ventricle'</td>\n",
       "      <td>b'GSE49906'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>118366262</td>\n",
       "      <td>100691769</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>RNeasy fibrous tissue mini kit was used for ve...</td>\n",
       "      <td>strain: FvB\\tgenotype/variation: wild type\\tag...</td>\n",
       "      <td>TopHat 2.0.5 is used to align reads to referen...</td>\n",
       "      <td>GSM1223636</td>\n",
       "      <td>SRA</td>\n",
       "      <td>RNA-seq ventricle PN28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Oct 20 2016'</td>\n",
       "      <td>b'Dendritic cells (DC)'</td>\n",
       "      <td>b'GSE88998'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>44508037</td>\n",
       "      <td>22091706</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was prepared from frozen purified ce...</td>\n",
       "      <td>strain: C57BL/6\\tSex: male\\ttissue: Pooled ski...</td>\n",
       "      <td>Paired-end raw read FASTq files ere pre-proces...</td>\n",
       "      <td>GSM2356925</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Untreated, CD11b, Replicate 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Feb 24 2016'</td>\n",
       "      <td>b'Cortex'</td>\n",
       "      <td>b'GSE78272'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>41451923</td>\n",
       "      <td>34162109</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Tissue specimens were removed, flash frozen on...</td>\n",
       "      <td>strain: C57BL/6\\ttissue: Cortex\\tgenotype: Wil...</td>\n",
       "      <td>Illumina CASAVA 1.8.2 software used for baseca...</td>\n",
       "      <td>GSM2070846</td>\n",
       "      <td>SRA</td>\n",
       "      <td>WT 10 months Male Cortex mRNA rep4 399031_8499...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jul 27 2016'</td>\n",
       "      <td>b'embryo'</td>\n",
       "      <td>b'GSE74155'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1702780</td>\n",
       "      <td>1321388</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Cells are directly picked after dissociation a...</td>\n",
       "      <td>strain: male CAST/EiJ x female C57BL/6J\\tcell ...</td>\n",
       "      <td>Illumina Casava1.7 software used for basecalli...</td>\n",
       "      <td>GSM2083218</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Epi344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 29 2014'</td>\n",
       "      <td>b'medial ganglionic eminence'</td>\n",
       "      <td>b'GSE60402'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>1895066</td>\n",
       "      <td>903476</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>E12.5 embryos were immersed in Tissue Tek (Sak...</td>\n",
       "      <td>animal: Wildtype2\\ttissue: medial ganglionic e...</td>\n",
       "      <td>Read quality filtering and barcode extraction:...</td>\n",
       "      <td>GSM1478010</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Wildtype2_5_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jan 30 2015'</td>\n",
       "      <td>b'Primary bone marrow-derived macrophages (BMDM)'</td>\n",
       "      <td>b'GSE56122\\tGSE56123'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>127650415</td>\n",
       "      <td>103088157</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>b'Mus musculus'</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was extracted from 5x106 cells using...</td>\n",
       "      <td>strain: BXH2/TyJ\\tcell type: BMDM cells (7th d...</td>\n",
       "      <td>Reads were quality filtered according to the I...</td>\n",
       "      <td>GSM1356226</td>\n",
       "      <td>SRA</td>\n",
       "      <td>BXH2_LPS_UT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0 taxid_ch1                    status  \\\n",
       "0            0  b'10090'  b'Public on Apr 07 2015'   \n",
       "1            1  b'10090'  b'Public on Jun 10 2016'   \n",
       "2            2  b'10090'  b'Public on Jun 10 2016'   \n",
       "3            3  b'10090'  b'Public on Jan 01 2016'   \n",
       "4            4  b'10090'  b'Public on Sep 10 2012'   \n",
       "5            5  b'10090'  b'Public on Oct 27 2016'   \n",
       "6            6  b'10090'  b'Public on Jun 10 2016'   \n",
       "7            7  b'10090'  b'Public on Feb 20 2015'   \n",
       "8            8  b'10090'  b'Public on Oct 06 2016'   \n",
       "9            9  b'10090'  b'Public on Jun 10 2016'   \n",
       "10          10  b'10090'  b'Public on Feb 20 2015'   \n",
       "11          11  b'10090'  b'Public on Jun 10 2016'   \n",
       "12          12  b'10090'  b'Public on Feb 20 2015'   \n",
       "13          13  b'10090'  b'Public on Apr 10 2013'   \n",
       "14          14  b'10090'  b'Public on Apr 16 2014'   \n",
       "15          15  b'10090'  b'Public on Oct 20 2016'   \n",
       "16          16  b'10090'  b'Public on Feb 24 2016'   \n",
       "17          17  b'10090'  b'Public on Jul 27 2016'   \n",
       "18          18  b'10090'  b'Public on Sep 29 2014'   \n",
       "19          19  b'10090'  b'Public on Jan 30 2015'   \n",
       "\n",
       "                                      source_name_ch1  \\\n",
       "0                                    b'kidney cortex'   \n",
       "1                                  b'corpus callosum'   \n",
       "2                                      b'dorsal horn'   \n",
       "3                                      b'Lung tissue'   \n",
       "4                                           b'Cortex'   \n",
       "5                                b'E12.5 Single Cell'   \n",
       "6                                  b'corpus callosum'   \n",
       "7                                         b'sscortex'   \n",
       "8                                 b'ventral midbrain'   \n",
       "9                                           b'SN-VTA'   \n",
       "10                                  b'ca1hippocampus'   \n",
       "11                                          b'SN-VTA'   \n",
       "12                                        b'sscortex'   \n",
       "13                  b'Dicer -/- murine MSC cell line'   \n",
       "14                                       b'ventricle'   \n",
       "15                            b'Dendritic cells (DC)'   \n",
       "16                                          b'Cortex'   \n",
       "17                                          b'embryo'   \n",
       "18                      b'medial ganglionic eminence'   \n",
       "19  b'Primary bone marrow-derived macrophages (BMDM)'   \n",
       "\n",
       "                          series_id  \\\n",
       "0                       b'GSE54390'   \n",
       "1                       b'GSE75330'   \n",
       "2                       b'GSE75330'   \n",
       "3                       b'GSE64027'   \n",
       "4   b'GSE36025\\tGSE49417\\tGSE49847'   \n",
       "5                       b'GSE87069'   \n",
       "6                       b'GSE75330'   \n",
       "7                       b'GSE60361'   \n",
       "8                       b'GSE76381'   \n",
       "9                       b'GSE75330'   \n",
       "10                      b'GSE60361'   \n",
       "11                      b'GSE75330'   \n",
       "12                      b'GSE60361'   \n",
       "13            b'GSE44157\\tGSE44163'   \n",
       "14                      b'GSE49906'   \n",
       "15                      b'GSE88998'   \n",
       "16                      b'GSE78272'   \n",
       "17                      b'GSE74155'   \n",
       "18                      b'GSE60402'   \n",
       "19            b'GSE56122\\tGSE56123'   \n",
       "\n",
       "                                             relation  readstotal  \\\n",
       "0   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    39116064   \n",
       "1   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1136541   \n",
       "2   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      514285   \n",
       "3   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    26804780   \n",
       "4   b'Reanalyzed by: GSE80797\\tSuperseded by: GSE9...   265978941   \n",
       "5   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     5607362   \n",
       "6   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1065010   \n",
       "7   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1211044   \n",
       "8   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      663629   \n",
       "9   b'BioSample: https://www.ncbi.nlm.nih.gov/bios...      456330   \n",
       "10  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     2534097   \n",
       "11  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1362059   \n",
       "12  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     3861394   \n",
       "13  b'Reanalyzed by: GSE80797\\tSRA: https://www.nc...    27584764   \n",
       "14  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...   118366262   \n",
       "15  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    44508037   \n",
       "16  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    41451923   \n",
       "17  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1702780   \n",
       "18  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...     1895066   \n",
       "19  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...   127650415   \n",
       "\n",
       "    readsaligned platform_id     organism_ch1  ... library_strategy.1  \\\n",
       "0       32884497    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "1         308433    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "2         148668    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "3       23355974    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "4      173619445    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "5        1929834    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "6         473113    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "7         330539    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "8         133904    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "9         131980    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "10        584659    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "11        570834    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "12       1464176    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "13      25018842    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "14     100691769    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "15      22091706    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "16      34162109    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "17       1321388    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "18        903476    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "19     103088157    GPL13112  b'Mus musculus'  ...            RNA-Seq   \n",
       "\n",
       "   library_selection.1 library_source.1   instrument_model.1  \\\n",
       "0                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "1                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "2                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "3                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "4                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "5                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "6                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "7                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "8                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "9                 cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "10                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "11                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "12                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "13                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "14                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "15                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "16                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "17                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "18                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "19                cDNA   transcriptomic  Illumina HiSeq 2000   \n",
       "\n",
       "                               extract_protocol_ch1.1  \\\n",
       "0   Thin outer slice of kidney cortex was immediat...   \n",
       "1   Selected CNS regions were dissected in 300µm t...   \n",
       "2   Selected CNS regions were dissected in 300µm t...   \n",
       "3   Total RNA was extracted from pooled dissected ...   \n",
       "4   http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...   \n",
       "5   Smart-seq2\\tNextera XT DNA library preparation...   \n",
       "6   Selected CNS regions were dissected in 300µm t...   \n",
       "7   Mice were deeply anesthetized with ketamine/xy...   \n",
       "8   Human embryo ventral midbrain dissection was p...   \n",
       "9   Selected CNS regions were dissected in 300µm t...   \n",
       "10  Mice were deeply anesthetized with ketamine/xy...   \n",
       "11  Selected CNS regions were dissected in 300µm t...   \n",
       "12  Mice were deeply anesthetized with ketamine/xy...   \n",
       "13  mRNA-sequencing was carried out with total RNA...   \n",
       "14  RNeasy fibrous tissue mini kit was used for ve...   \n",
       "15  Total RNA was prepared from frozen purified ce...   \n",
       "16  Tissue specimens were removed, flash frozen on...   \n",
       "17  Cells are directly picked after dissociation a...   \n",
       "18  E12.5 embryos were immersed in Tissue Tek (Sak...   \n",
       "19  Total RNA was extracted from 5x106 cells using...   \n",
       "\n",
       "                                characteristics_ch1.1  \\\n",
       "0   tissue: kidney cortex\\tstrain: C57BL/6\\tgender...   \n",
       "1   strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...   \n",
       "2   strain: C57BL6\\tage: p24\\tSex: ?\\ttreatment: N...   \n",
       "3   strain: A/J\\tSex: female\\ttissue: Lung\\ttime p...   \n",
       "4   lab: CSHL-m\\tlab description: Gingeras - Cold ...   \n",
       "5   strain: C57BL/6 Heterozygous Lmx1a-EGFP\\tdevel...   \n",
       "6   strain: PDGFRa-Cre\\tage: p61\\tSex: M\\ttreatmen...   \n",
       "7                                 Sex: male\\tage: P23   \n",
       "8   tissue: ventral midbrain\\tSex: pooled male and...   \n",
       "9   strain: CD1\\tage: p20\\tSex: pooled male and fe...   \n",
       "10                                Sex: male\\tage: P23   \n",
       "11  strain: CD1\\tage: p21\\tSex: pooled male and fe...   \n",
       "12                                Sex: male\\tage: P26   \n",
       "13  cell type: murine MSC cell line\\tgenotype: Dic...   \n",
       "14  strain: FvB\\tgenotype/variation: wild type\\tag...   \n",
       "15  strain: C57BL/6\\tSex: male\\ttissue: Pooled ski...   \n",
       "16  strain: C57BL/6\\ttissue: Cortex\\tgenotype: Wil...   \n",
       "17  strain: male CAST/EiJ x female C57BL/6J\\tcell ...   \n",
       "18  animal: Wildtype2\\ttissue: medial ganglionic e...   \n",
       "19  strain: BXH2/TyJ\\tcell type: BMDM cells (7th d...   \n",
       "\n",
       "                                    data_processing.1 geo_accession.1 type.1  \\\n",
       "0   Reads were aligned to the genome using the Bur...      GSM1314260    SRA   \n",
       "1   Read processing was performed as described (Is...      GSM2164318    SRA   \n",
       "2   Read processing was performed as described (Is...      GSM2161048    SRA   \n",
       "3   Illumina Casava1.8 software used for basecalli...      GSM1563082    SRA   \n",
       "4   http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm...      GSM1000563    SRA   \n",
       "5   reads were aligned to the mm10 genome using de...      GSM2320580    SRA   \n",
       "6   Read processing was performed as described (Is...      GSM2164283    SRA   \n",
       "7   Read processing was performed as described (Is...      GSM1474588    SRA   \n",
       "8   Read processing was performed as described (Is...      GSM2273393    SRA   \n",
       "9   Read processing was performed as described (Is...      GSM2162146    SRA   \n",
       "10  Read processing was performed as described (Is...      GSM1475854    SRA   \n",
       "11  Read processing was performed as described (Is...      GSM2161497    SRA   \n",
       "12  Read processing was performed as described (Is...      GSM1474327    SRA   \n",
       "13  Illumina pipeline version 1.3+ used for baseca...      GSM1079794    SRA   \n",
       "14  TopHat 2.0.5 is used to align reads to referen...      GSM1223636    SRA   \n",
       "15  Paired-end raw read FASTq files ere pre-proces...      GSM2356925    SRA   \n",
       "16  Illumina CASAVA 1.8.2 software used for baseca...      GSM2070846    SRA   \n",
       "17  Illumina Casava1.7 software used for basecalli...      GSM2083218    SRA   \n",
       "18  Read quality filtering and barcode extraction:...      GSM1478010    SRA   \n",
       "19  Reads were quality filtered according to the I...      GSM1356226    SRA   \n",
       "\n",
       "                                              title.1  \n",
       "0                                TRACK 43-75 kidney 1  \n",
       "1                                  C1-1772117-029-A02  \n",
       "2                                  C1-1772072-242-F03  \n",
       "3                                  20NNK_LPS_Sil_I3C9  \n",
       "4   CSHL_RnaSeq_Cortex_adult-8wks (superseded by G...  \n",
       "5                                        F125_RNA-seq  \n",
       "6                                  C1-1772117-028-A05  \n",
       "7                                     SingleCellNo338  \n",
       "8                                      1772075301_G03  \n",
       "9                                  C1-1772096-087-D08  \n",
       "10                                   SingleCellNo1599  \n",
       "11                                 C1-1772089-200-B02  \n",
       "12                                     SingleCellNo77  \n",
       "13                                      KO_siNr6a1_R0  \n",
       "14                             RNA-seq ventricle PN28  \n",
       "15                      Untreated, CD11b, Replicate 3  \n",
       "16  WT 10 months Male Cortex mRNA rep4 399031_8499...  \n",
       "17                                             Epi344  \n",
       "18                                      Wildtype2_5_3  \n",
       "19                                        BXH2_LPS_UT  \n",
       "\n",
       "[20 rows x 41 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "fdadae01-7345-41b6-ae3e-953f48259ed5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>status</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>series_id</th>\n",
       "      <th>relation</th>\n",
       "      <th>readstotal</th>\n",
       "      <th>readsaligned</th>\n",
       "      <th>platform_id</th>\n",
       "      <th>...</th>\n",
       "      <th>library_strategy.1</th>\n",
       "      <th>library_selection.1</th>\n",
       "      <th>library_source.1</th>\n",
       "      <th>instrument_model.1</th>\n",
       "      <th>extract_protocol_ch1.1</th>\n",
       "      <th>characteristics_ch1.1</th>\n",
       "      <th>data_processing.1</th>\n",
       "      <th>geo_accession.1</th>\n",
       "      <th>type.1</th>\n",
       "      <th>title.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jan 30 2015'</td>\n",
       "      <td>b'Primary bone marrow-derived macrophages (BMDM)'</td>\n",
       "      <td>b'GSE56122\\tGSE56123'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>127650415</td>\n",
       "      <td>103088157</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Total RNA was extracted from 5x106 cells using...</td>\n",
       "      <td>strain: BXH2/TyJ\\tcell type: BMDM cells (7th d...</td>\n",
       "      <td>Reads were quality filtered according to the I...</td>\n",
       "      <td>GSM1356226</td>\n",
       "      <td>SRA</td>\n",
       "      <td>BXH2_LPS_UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jul 15 2016'</td>\n",
       "      <td>b'Pro B cells'</td>\n",
       "      <td>b'GSE72018\\tGSE72019'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>50911147</td>\n",
       "      <td>45076587</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>Female mouse C57BL/6 B cell populations were s...</td>\n",
       "      <td>strain: C57BL/6JNimr\\ttissue: Bone marrow\\tage...</td>\n",
       "      <td>fastq files were created with CASAVA 1.8.1\\tSe...</td>\n",
       "      <td>GSM1849922</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Bcell-pro-R5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>39</td>\n",
       "      <td>39</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 20 2016'</td>\n",
       "      <td>b'Follicular B cells'</td>\n",
       "      <td>b'GSE73532\\tGSE73534'</td>\n",
       "      <td>b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...</td>\n",
       "      <td>12791738</td>\n",
       "      <td>4090516</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>For bone marrow subset purification we used B6...</td>\n",
       "      <td>strain: C57Bl/6\\tage: 6-14 weeks\\tgender: pool...</td>\n",
       "      <td>Raw data were aligned to the mouse reference g...</td>\n",
       "      <td>GSM1897413</td>\n",
       "      <td>SRA</td>\n",
       "      <td>FOB RNA-seq_r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Apr 07 2016'</td>\n",
       "      <td>b'Mouse CD4+ cells 4h after activation'</td>\n",
       "      <td>b'GSE52260'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>32010611</td>\n",
       "      <td>24778423</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>RNA from human and mouse samples was extracted...</td>\n",
       "      <td>source: C57BL/6\\tcell type: CD4+\\tmouse biolog...</td>\n",
       "      <td>Illumina Casava1.8 software used for basecalli...</td>\n",
       "      <td>GSM1262088</td>\n",
       "      <td>SRA</td>\n",
       "      <td>C57BL/6 Th0 4h rep3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77</td>\n",
       "      <td>77</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Jun 05 2013'</td>\n",
       "      <td>b'bone marrow derived macrophages'</td>\n",
       "      <td>b'GSE45914'</td>\n",
       "      <td>b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...</td>\n",
       "      <td>136512935</td>\n",
       "      <td>16000854</td>\n",
       "      <td>GPL13112</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>ChIP-seq: Lysates were clarified from sonicate...</td>\n",
       "      <td>cell type: bone marrow derived macrophages (Re...</td>\n",
       "      <td>For ChIP-Seq and GRO-Seq samples (single end r...</td>\n",
       "      <td>GSM1119597</td>\n",
       "      <td>SRA</td>\n",
       "      <td>RevErb WT BMDM, GRO-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9405</th>\n",
       "      <td>500928</td>\n",
       "      <td>500928</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Oct 07 2021'</td>\n",
       "      <td>b'macrophages_Dnmt3a KO'</td>\n",
       "      <td>b'GSE183480\\tGSE184446'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>18404806</td>\n",
       "      <td>16283855</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Total RNA from BMDMs were isolated using the R...</td>\n",
       "      <td>strain: C57BL/6\\tgenotype/variation: Dnmt3a KO...</td>\n",
       "      <td>Raw sequencing fastq files were trimmed with T...</td>\n",
       "      <td>GSM5558224</td>\n",
       "      <td>SRA</td>\n",
       "      <td>RNA-seq Dnmt3a KO rep1 [D_KO_BMDM_9]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9406</th>\n",
       "      <td>500929</td>\n",
       "      <td>500929</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Oct 07 2021'</td>\n",
       "      <td>b'macrophages_Dnmt3a KO'</td>\n",
       "      <td>b'GSE183480\\tGSE184446'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>17712338</td>\n",
       "      <td>15570776</td>\n",
       "      <td>GPL21626</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>Total RNA from BMDMs were isolated using the R...</td>\n",
       "      <td>strain: C57BL/6\\tgenotype/variation: Dnmt3a KO...</td>\n",
       "      <td>Raw sequencing fastq files were trimmed with T...</td>\n",
       "      <td>GSM5558226</td>\n",
       "      <td>SRA</td>\n",
       "      <td>RNA-seq Dnmt3a KO rep3 [D_KO_BMDM_11]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9407</th>\n",
       "      <td>500932</td>\n",
       "      <td>500932</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'Lung-tumor infiltrating CD4+ T cells'</td>\n",
       "      <td>b'GSE183544'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>25742942</td>\n",
       "      <td>23112530</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Lung tissues from CD4/ALK4-KO or WT mice beari...</td>\n",
       "      <td>tissue: Lung tumor\\tcell type: CD4+ T cells\\tg...</td>\n",
       "      <td>Quality Control was performed at the fastq raw...</td>\n",
       "      <td>GSM5560174</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Wild_Type_rep1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9408</th>\n",
       "      <td>500933</td>\n",
       "      <td>500933</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'Lung-tumor infiltrating CD4+ T cells'</td>\n",
       "      <td>b'GSE183544'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>29155148</td>\n",
       "      <td>26105830</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Lung tissues from CD4/ALK4-KO or WT mice beari...</td>\n",
       "      <td>tissue: Lung tumor\\tcell type: CD4+ T cells\\tg...</td>\n",
       "      <td>Quality Control was performed at the fastq raw...</td>\n",
       "      <td>GSM5560175</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Wild_Type_rep2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9409</th>\n",
       "      <td>500934</td>\n",
       "      <td>500934</td>\n",
       "      <td>b'10090'</td>\n",
       "      <td>b'Public on Sep 10 2021'</td>\n",
       "      <td>b'Lung-tumor infiltrating CD4+ T cells'</td>\n",
       "      <td>b'GSE183544'</td>\n",
       "      <td>b'BioSample: https://www.ncbi.nlm.nih.gov/bios...</td>\n",
       "      <td>21770352</td>\n",
       "      <td>19064430</td>\n",
       "      <td>GPL19057</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>transcriptomic</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>Lung tissues from CD4/ALK4-KO or WT mice beari...</td>\n",
       "      <td>tissue: Lung tumor\\tcell type: CD4+ T cells\\tg...</td>\n",
       "      <td>Quality Control was performed at the fastq raw...</td>\n",
       "      <td>GSM5560176</td>\n",
       "      <td>SRA</td>\n",
       "      <td>CD4/ALK4-KO_rep1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9410 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index  Unnamed: 0 taxid_ch1                    status  \\\n",
       "0         19          19  b'10090'  b'Public on Jan 30 2015'   \n",
       "1         28          28  b'10090'  b'Public on Jul 15 2016'   \n",
       "2         39          39  b'10090'  b'Public on Jun 20 2016'   \n",
       "3         65          65  b'10090'  b'Public on Apr 07 2016'   \n",
       "4         77          77  b'10090'  b'Public on Jun 05 2013'   \n",
       "...      ...         ...       ...                       ...   \n",
       "9405  500928      500928  b'10090'  b'Public on Oct 07 2021'   \n",
       "9406  500929      500929  b'10090'  b'Public on Oct 07 2021'   \n",
       "9407  500932      500932  b'10090'  b'Public on Sep 10 2021'   \n",
       "9408  500933      500933  b'10090'  b'Public on Sep 10 2021'   \n",
       "9409  500934      500934  b'10090'  b'Public on Sep 10 2021'   \n",
       "\n",
       "                                        source_name_ch1  \\\n",
       "0     b'Primary bone marrow-derived macrophages (BMDM)'   \n",
       "1                                        b'Pro B cells'   \n",
       "2                                 b'Follicular B cells'   \n",
       "3               b'Mouse CD4+ cells 4h after activation'   \n",
       "4                    b'bone marrow derived macrophages'   \n",
       "...                                                 ...   \n",
       "9405                           b'macrophages_Dnmt3a KO'   \n",
       "9406                           b'macrophages_Dnmt3a KO'   \n",
       "9407            b'Lung-tumor infiltrating CD4+ T cells'   \n",
       "9408            b'Lung-tumor infiltrating CD4+ T cells'   \n",
       "9409            b'Lung-tumor infiltrating CD4+ T cells'   \n",
       "\n",
       "                    series_id  \\\n",
       "0       b'GSE56122\\tGSE56123'   \n",
       "1       b'GSE72018\\tGSE72019'   \n",
       "2       b'GSE73532\\tGSE73534'   \n",
       "3                 b'GSE52260'   \n",
       "4                 b'GSE45914'   \n",
       "...                       ...   \n",
       "9405  b'GSE183480\\tGSE184446'   \n",
       "9406  b'GSE183480\\tGSE184446'   \n",
       "9407             b'GSE183544'   \n",
       "9408             b'GSE183544'   \n",
       "9409             b'GSE183544'   \n",
       "\n",
       "                                               relation  readstotal  \\\n",
       "0     b'BioSample: https://www.ncbi.nlm.nih.gov/bios...   127650415   \n",
       "1     b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    50911147   \n",
       "2     b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...    12791738   \n",
       "3     b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    32010611   \n",
       "4     b'SRA: https://www.ncbi.nlm.nih.gov/sra?term=S...   136512935   \n",
       "...                                                 ...         ...   \n",
       "9405  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    18404806   \n",
       "9406  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    17712338   \n",
       "9407  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    25742942   \n",
       "9408  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    29155148   \n",
       "9409  b'BioSample: https://www.ncbi.nlm.nih.gov/bios...    21770352   \n",
       "\n",
       "      readsaligned platform_id  ... library_strategy.1 library_selection.1  \\\n",
       "0        103088157    GPL13112  ...            RNA-Seq                cDNA   \n",
       "1         45076587    GPL13112  ...            RNA-Seq                cDNA   \n",
       "2          4090516    GPL13112  ...            RNA-Seq                cDNA   \n",
       "3         24778423    GPL13112  ...            RNA-Seq                cDNA   \n",
       "4         16000854    GPL13112  ...            RNA-Seq                cDNA   \n",
       "...            ...         ...  ...                ...                 ...   \n",
       "9405      16283855    GPL21626  ...            RNA-Seq                cDNA   \n",
       "9406      15570776    GPL21626  ...            RNA-Seq                cDNA   \n",
       "9407      23112530    GPL19057  ...            RNA-Seq                cDNA   \n",
       "9408      26105830    GPL19057  ...            RNA-Seq                cDNA   \n",
       "9409      19064430    GPL19057  ...            RNA-Seq                cDNA   \n",
       "\n",
       "     library_source.1    instrument_model.1  \\\n",
       "0      transcriptomic   Illumina HiSeq 2000   \n",
       "1      transcriptomic   Illumina HiSeq 2000   \n",
       "2      transcriptomic   Illumina HiSeq 2000   \n",
       "3      transcriptomic   Illumina HiSeq 2000   \n",
       "4      transcriptomic   Illumina HiSeq 2000   \n",
       "...               ...                   ...   \n",
       "9405   transcriptomic           NextSeq 550   \n",
       "9406   transcriptomic           NextSeq 550   \n",
       "9407   transcriptomic  Illumina NextSeq 500   \n",
       "9408   transcriptomic  Illumina NextSeq 500   \n",
       "9409   transcriptomic  Illumina NextSeq 500   \n",
       "\n",
       "                                 extract_protocol_ch1.1  \\\n",
       "0     Total RNA was extracted from 5x106 cells using...   \n",
       "1     Female mouse C57BL/6 B cell populations were s...   \n",
       "2     For bone marrow subset purification we used B6...   \n",
       "3     RNA from human and mouse samples was extracted...   \n",
       "4     ChIP-seq: Lysates were clarified from sonicate...   \n",
       "...                                                 ...   \n",
       "9405  Total RNA from BMDMs were isolated using the R...   \n",
       "9406  Total RNA from BMDMs were isolated using the R...   \n",
       "9407  Lung tissues from CD4/ALK4-KO or WT mice beari...   \n",
       "9408  Lung tissues from CD4/ALK4-KO or WT mice beari...   \n",
       "9409  Lung tissues from CD4/ALK4-KO or WT mice beari...   \n",
       "\n",
       "                                  characteristics_ch1.1  \\\n",
       "0     strain: BXH2/TyJ\\tcell type: BMDM cells (7th d...   \n",
       "1     strain: C57BL/6JNimr\\ttissue: Bone marrow\\tage...   \n",
       "2     strain: C57Bl/6\\tage: 6-14 weeks\\tgender: pool...   \n",
       "3     source: C57BL/6\\tcell type: CD4+\\tmouse biolog...   \n",
       "4     cell type: bone marrow derived macrophages (Re...   \n",
       "...                                                 ...   \n",
       "9405  strain: C57BL/6\\tgenotype/variation: Dnmt3a KO...   \n",
       "9406  strain: C57BL/6\\tgenotype/variation: Dnmt3a KO...   \n",
       "9407  tissue: Lung tumor\\tcell type: CD4+ T cells\\tg...   \n",
       "9408  tissue: Lung tumor\\tcell type: CD4+ T cells\\tg...   \n",
       "9409  tissue: Lung tumor\\tcell type: CD4+ T cells\\tg...   \n",
       "\n",
       "                                      data_processing.1 geo_accession.1  \\\n",
       "0     Reads were quality filtered according to the I...      GSM1356226   \n",
       "1     fastq files were created with CASAVA 1.8.1\\tSe...      GSM1849922   \n",
       "2     Raw data were aligned to the mouse reference g...      GSM1897413   \n",
       "3     Illumina Casava1.8 software used for basecalli...      GSM1262088   \n",
       "4     For ChIP-Seq and GRO-Seq samples (single end r...      GSM1119597   \n",
       "...                                                 ...             ...   \n",
       "9405  Raw sequencing fastq files were trimmed with T...      GSM5558224   \n",
       "9406  Raw sequencing fastq files were trimmed with T...      GSM5558226   \n",
       "9407  Quality Control was performed at the fastq raw...      GSM5560174   \n",
       "9408  Quality Control was performed at the fastq raw...      GSM5560175   \n",
       "9409  Quality Control was performed at the fastq raw...      GSM5560176   \n",
       "\n",
       "     type.1                                title.1  \n",
       "0       SRA                            BXH2_LPS_UT  \n",
       "1       SRA                           Bcell-pro-R5  \n",
       "2       SRA                         FOB RNA-seq_r1  \n",
       "3       SRA                    C57BL/6 Th0 4h rep3  \n",
       "4       SRA                RevErb WT BMDM, GRO-seq  \n",
       "...     ...                                    ...  \n",
       "9405    SRA   RNA-seq Dnmt3a KO rep1 [D_KO_BMDM_9]  \n",
       "9406    SRA  RNA-seq Dnmt3a KO rep3 [D_KO_BMDM_11]  \n",
       "9407    SRA                         Wild_Type_rep1  \n",
       "9408    SRA                         Wild_Type_rep2  \n",
       "9409    SRA                       CD4/ALK4-KO_rep1  \n",
       "\n",
       "[9410 rows x 42 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "abc4e699-abbb-41a0-9f03-f30871194745",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0           19\n",
       "1           28\n",
       "2           39\n",
       "3           65\n",
       "4           77\n",
       "         ...  \n",
       "9405    500928\n",
       "9406    500929\n",
       "9407    500932\n",
       "9408    500933\n",
       "9409    500934\n",
       "Name: index, Length: 9410, dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "94468aca-ad04-4e96-8fe6-68ff9d7f19a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[19, 28, 39, 65, 77, 107, 184, 241, 275, 280]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "indices2[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "137b8fa5-9191-4bca-a7a0-8ce9e84a3379",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9410"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(indices2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "06b5520c-b129-4891-beb6-fd0ff3e6eada",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "501054"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "3f1e8178-fd7e-4e8b-b632-1ad850971cc3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b'GSM1356226'\n",
      "501054\n",
      "(501054,)\n"
     ]
    }
   ],
   "source": [
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "    data_set=f['meta']['samples']['geo_accession']\n",
    "    print(data_set[19])\n",
    "    print(len(data_set))\n",
    "    print(data_set.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "05042462-9c5a-458d-bcc2-1f44abdce26e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2 0 0 ... 0 1 2]\n",
      "32544\n",
      "2\n",
      "501054\n",
      "[  3228   1185    114 ...  22074 137790  55081]\n"
     ]
    }
   ],
   "source": [
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "    data_set=f['data']['expression']\n",
    "    print(data_set[19])\n",
    "    print(len(data_set))\n",
    "    print(len(data_set.shape))\n",
    "    print(len(data_set[19]))\n",
    "    print(data_set[:, 19])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a441211-516b-4722-818c-608bf2ffa3ed",
   "metadata": {},
   "source": [
    "create an expression table\n",
    "'''\n",
    "'''\n",
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "   data_set = f['data']['expression']\n",
    "   data=data_set[::,indices2]\n",
    "   expression = pd.DataFrame(data, columns=indices2)\n",
    "   expression.to_csv('expression.csv')\n",
    "   f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9da6a342-937f-4fa0-9b15-351408ce42d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "    data_set = f['data']['expression']\n",
    "    data=data_set[::,indices2]\n",
    "    expression = pd.DataFrame(data, columns=indices)\n",
    "    expression.to_csv('expression3.csv')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2674247b-34bd-476a-b45d-a26717581e9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32544\n",
      "9410\n"
     ]
    }
   ],
   "source": [
    "print(len(expression))\n",
    "print(expression.shape[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "733d26e0-6b6c-4163-803f-9c9004610bc7",
   "metadata": {},
   "source": [
    "Заменим индексы колонок в таблице экспрессии на gsm\n",
    "GSM1849922 - должно быть в позиции 28 - и это верно"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bd6ed0cd-4e50-4b8e-89ab-3dae76c23578",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/xm/lz8thrrs1l5cks282q243ty40000gn/T/ipykernel_2926/184194338.py:1: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  wf=pd.read_csv('table.csv')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "geo_accession  GSM1356226  GSM1849922  GSM1897413  GSM1262088  GSM1119597  \\\n",
      "0                    3228        1274         114         810         440   \n",
      "1                    1185         437          12          56         151   \n",
      "2                     114          35          21          11         138   \n",
      "3                    3548        3126          95         520         254   \n",
      "4                    4075        1033          40         157         465   \n",
      "\n",
      "geo_accession  GSM1533925  GSM1511235  GSM1808311  GSM1823626  GSM1713923  \\\n",
      "0                     117         500        1425         326        1121   \n",
      "1                      78         136         757         237         200   \n",
      "2                       4          46         101          15          44   \n",
      "3                     112        1954        1063         558        1140   \n",
      "4                     261          54        1834         948         441   \n",
      "\n",
      "geo_accession  ...  GSM5550906  GSM5550908  GSM5550909  GSM5558220  \\\n",
      "0              ...        1629        1079         839         372   \n",
      "1              ...        1570         966        1878         328   \n",
      "2              ...         155          72         107          44   \n",
      "3              ...        1304         882         928         395   \n",
      "4              ...        3095        1666        2484         494   \n",
      "\n",
      "geo_accession  GSM5558221  GSM5558224  GSM5558226  GSM5560174  GSM5560175  \\\n",
      "0                     446         380         364         926         799   \n",
      "1                     295         245         251         301         327   \n",
      "2                      24          24          17           5           3   \n",
      "3                     372         278         253         265         230   \n",
      "4                     838         600         709         726         751   \n",
      "\n",
      "geo_accession  GSM5560176  \n",
      "0                     562  \n",
      "1                     153  \n",
      "2                       9  \n",
      "3                     105  \n",
      "4                     581  \n",
      "\n",
      "[5 rows x 9410 columns]\n"
     ]
    }
   ],
   "source": [
    "wf=pd.read_csv('table.csv')\n",
    "#df.drop(df.columns[[0]], axis = 1)\n",
    "expression.columns = wf.loc[indices]['geo_accession']\n",
    "print(expression.head())\n",
    "expression.to_csv('expression4.csv')   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "4428c0a2-00ab-4792-bb96-e652d6fe646d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    19\n",
      "1    28\n",
      "2    39\n",
      "3    65\n",
      "4    77\n",
      "Name: index, dtype: int64\n",
      "[19, 28, 39, 65, 77]\n"
     ]
    }
   ],
   "source": [
    "print(indices.head())\n",
    "print(indices2[0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "805d4e0e-e4dc-4662-8a23-3a4430ea42f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unnamed: 0                                                             28\n",
      "taxid_ch1                                                        b'10090'\n",
      "status                                           b'Public on Jul 15 2016'\n",
      "source_name_ch1                                            b'Pro B cells'\n",
      "series_id                                           b'GSE72018\\tGSE72019'\n",
      "relation                b'BioSample: https://www.ncbi.nlm.nih.gov/bios...\n",
      "readstotal                                                       50911147\n",
      "readsaligned                                                     45076587\n",
      "platform_id                                                      GPL13112\n",
      "organism_ch1                                              b'Mus musculus'\n",
      "molecule_ch1                                                 b'total RNA'\n",
      "library_strategy                                               b'RNA-Seq'\n",
      "library_selection                                                 b'cDNA'\n",
      "library_source                                          b'transcriptomic'\n",
      "instrument_model                                   b'Illumina HiSeq 2000'\n",
      "extract_protocol_ch1    b\"Female mouse C57BL/6 B cell populations were...\n",
      "characteristics_ch1     b'strain: C57BL/6JNimr\\ttissue: Bone marrow\\ta...\n",
      "data_processing         b'fastq files were created with CASAVA 1.8.1\\t...\n",
      "geo_accession                                                  GSM1849922\n",
      "type                                                               b'SRA'\n",
      "title                                                     b'Bcell-pro-R5'\n",
      "Name: 28, dtype: object\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('table.csv')  \n",
    "print(df.iloc[28])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2c2eb8c-23f1-4cd8-9e1e-f27d5fcc60a1",
   "metadata": {},
   "source": [
    "Теперь строки в таблице с экспрессией номер 4 нужно заменить названия строк на гены - indices 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e1c7d8e1-e4bd-4281-8666-0f9335d063ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               0\n",
      "0  0610007P14Rik\n",
      "1  0610009B22Rik\n",
      "2  0610009L18Rik\n",
      "3  0610009O20Rik\n",
      "4  0610010F05Rik\n",
      "                    0\n",
      "0             missing\n",
      "1  ENSMUSG00000007777\n",
      "2  ENSMUSG00000043644\n",
      "3             missing\n",
      "4  ENSMUSG00000042208\n",
      "32544\n",
      "32544\n"
     ]
    }
   ],
   "source": [
    "with h5py.File('mouse_matrix_v11.h5', 'r') as f:\n",
    "    data_set = f['meta']['genes']['genes']\n",
    "    data=data_set[::]\n",
    "    names_lines = pd.DataFrame(data)\n",
    "    ensembl_set=f['meta']['genes']['ensembl_gene_id']\n",
    "    ensembl=ensembl_set[::]\n",
    "    ensemble = pd.DataFrame(ensembl) \n",
    "    for col, dtype in names_lines.dtypes.items():\n",
    "        if dtype == np.object_:  # Only process byte object columns.\n",
    "            names_lines[col] = names_lines[col].apply(lambda x: x.decode(\"utf-8\"))\n",
    "    for col, dtype in ensemble.dtypes.items():\n",
    "        if dtype == np.object_:  # Only process byte object columns.\n",
    "            ensemble[col] = ensemble[col].apply(lambda x: x.decode(\"utf-8\"))\n",
    "    f.close()\n",
    "print(names_lines.head())\n",
    "print(ensemble.head())\n",
    "print(len(ensemble))\n",
    "print(len(names_lines))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "51c49ab1-d483-4145-9471-d097af3c3ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    0              0  GSM1356226  GSM1849922  GSM1897413  \\\n",
      "0             missing  0610007P14Rik        3228        1274         114   \n",
      "1  ENSMUSG00000007777  0610009B22Rik        1185         437          12   \n",
      "2  ENSMUSG00000043644  0610009L18Rik         114          35          21   \n",
      "3             missing  0610009O20Rik        3548        3126          95   \n",
      "4  ENSMUSG00000042208  0610010F05Rik        4075        1033          40   \n",
      "\n",
      "   GSM1262088  GSM1119597  GSM1533925  GSM1511235  GSM1808311  ...  \\\n",
      "0         810         440         117         500        1425  ...   \n",
      "1          56         151          78         136         757  ...   \n",
      "2          11         138           4          46         101  ...   \n",
      "3         520         254         112        1954        1063  ...   \n",
      "4         157         465         261          54        1834  ...   \n",
      "\n",
      "   GSM5550906  GSM5550908  GSM5550909  GSM5558220  GSM5558221  GSM5558224  \\\n",
      "0        1629        1079         839         372         446         380   \n",
      "1        1570         966        1878         328         295         245   \n",
      "2         155          72         107          44          24          24   \n",
      "3        1304         882         928         395         372         278   \n",
      "4        3095        1666        2484         494         838         600   \n",
      "\n",
      "   GSM5558226  GSM5560174  GSM5560175  GSM5560176  \n",
      "0         364         926         799         562  \n",
      "1         251         301         327         153  \n",
      "2          17           5           3           9  \n",
      "3         253         265         230         105  \n",
      "4         709         726         751         581  \n",
      "\n",
      "[5 rows x 9412 columns]\n"
     ]
    }
   ],
   "source": [
    "expression=pd.read_csv('expression4.csv')\n",
    "expression2 = pd.concat([ensemble, names_lines, expression], axis=1)\n",
    "expression2.pop('Unnamed: 0')\n",
    "expression2.to_csv('expression5.csv')   \n",
    "print(expression2.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ce4d5e56-1aa1-460e-aca7-1621b6153e64",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = pd.read_csv('expression5.csv')  \n",
    "f.pop('Unnamed: 0')\n",
    "f.rename(columns={'0': \"ensemble\", '0.1': \"symbol\"}, inplace=True)\n",
    "f.to_csv('expression6.csv')   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "84bc80fa-beb4-4d81-bd47-8347aa5b4c64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             ensemble         symbol  GSM1356226  GSM1849922  GSM1897413  \\\n",
      "0             missing  0610007P14Rik        3228        1274         114   \n",
      "1  ENSMUSG00000007777  0610009B22Rik        1185         437          12   \n",
      "2  ENSMUSG00000043644  0610009L18Rik         114          35          21   \n",
      "3             missing  0610009O20Rik        3548        3126          95   \n",
      "4  ENSMUSG00000042208  0610010F05Rik        4075        1033          40   \n",
      "\n",
      "   GSM1262088  GSM1119597  GSM1533925  GSM1511235  GSM1808311  ...  \\\n",
      "0         810         440         117         500        1425  ...   \n",
      "1          56         151          78         136         757  ...   \n",
      "2          11         138           4          46         101  ...   \n",
      "3         520         254         112        1954        1063  ...   \n",
      "4         157         465         261          54        1834  ...   \n",
      "\n",
      "   GSM5550906  GSM5550908  GSM5550909  GSM5558220  GSM5558221  GSM5558224  \\\n",
      "0        1629        1079         839         372         446         380   \n",
      "1        1570         966        1878         328         295         245   \n",
      "2         155          72         107          44          24          24   \n",
      "3        1304         882         928         395         372         278   \n",
      "4        3095        1666        2484         494         838         600   \n",
      "\n",
      "   GSM5558226  GSM5560174  GSM5560175  GSM5560176  \n",
      "0         364         926         799         562  \n",
      "1         251         301         327         153  \n",
      "2          17           5           3           9  \n",
      "3         253         265         230         105  \n",
      "4         709         726         751         581  \n",
      "\n",
      "[5 rows x 9412 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(f.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad089ab-7692-4cc4-9e3a-4c692d0c51f2",
   "metadata": {},
   "source": [
    " ### Создать таблицу - перечение с той что пришлют"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b36504da-9b06-4c88-acc3-4ec528df1420",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   GSM4042993  GSM4042994  GSM4042995  GSM4042996  GSM4042997  GSM4042998  \\\n",
      "0           0           0           0           0           0           0   \n",
      "1           0           0           0           0           0           0   \n",
      "2           0           0           0           0           0           0   \n",
      "3           0           0           0           0           0           0   \n",
      "4           0           0           0           0           0           0   \n",
      "\n",
      "   GSM4042999  GSM4043000  GSM4043001  GSM4043002  ...  GSM1234735  \\\n",
      "0           0           0           0           0  ...           0   \n",
      "1           0           0           0           0  ...           1   \n",
      "2           0           0           0           0  ...           0   \n",
      "3           0           1           0           0  ...           0   \n",
      "4           0           0           0           0  ...           0   \n",
      "\n",
      "   GSM1234736  GSM1234737  GSM1234738  GSM1234739  GSM1234740  GSM1234741  \\\n",
      "0           0           0           0           0           0           0   \n",
      "1           0           0           0           0           0           0   \n",
      "2           6           0           1           0           0           2   \n",
      "3           0           0           0           0           0           0   \n",
      "4           0           0           1           0           0           0   \n",
      "\n",
      "   GSM1234742  GSM1234743  GSM1234744  \n",
      "0           0           0           0  \n",
      "1           0           1           0  \n",
      "2           0           0           3  \n",
      "3           0           0           0  \n",
      "4           0           0           1  \n",
      "\n",
      "[5 rows x 5558 columns]\n"
     ]
    }
   ],
   "source": [
    "dee2_exp=pd.read_csv('dee2_exp.csv')\n",
    "print(dee2_exp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "59831d09-916b-4da6-b353-40262a2565bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             ensemble         symbol\n",
      "0  ENSMUSG00000102693  4933401J01Rik\n",
      "1  ENSMUSG00000064842        Gm26206\n",
      "2  ENSMUSG00000051951           Xkr4\n",
      "3  ENSMUSG00000102851        Gm18956\n",
      "4  ENSMUSG00000103377        Gm37180\n",
      "52636\n",
      "5558\n"
     ]
    }
   ],
   "source": [
    "dee2_genes=pd.read_csv('dee2_genes.csv')\n",
    "print(dee2_genes.head())\n",
    "print(dee2_genes.shape[0])\n",
    "print(dee2_exp.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4593db8f-9c86-443d-aec3-a066866e8cd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             ensemble         symbol  GSM4042993  GSM4042994  GSM4042995  \\\n",
      "0  ENSMUSG00000102693  4933401J01Rik           0           0           0   \n",
      "1  ENSMUSG00000064842        Gm26206           0           0           0   \n",
      "2  ENSMUSG00000051951           Xkr4           0           0           0   \n",
      "3  ENSMUSG00000102851        Gm18956           0           0           0   \n",
      "4  ENSMUSG00000103377        Gm37180           0           0           0   \n",
      "\n",
      "   GSM4042996  GSM4042997  GSM4042998  GSM4042999  GSM4043000  ...  \\\n",
      "0           0           0           0           0           0  ...   \n",
      "1           0           0           0           0           0  ...   \n",
      "2           0           0           0           0           0  ...   \n",
      "3           0           0           0           0           1  ...   \n",
      "4           0           0           0           0           0  ...   \n",
      "\n",
      "   GSM1234735  GSM1234736  GSM1234737  GSM1234738  GSM1234739  GSM1234740  \\\n",
      "0           0           0           0           0           0           0   \n",
      "1           1           0           0           0           0           0   \n",
      "2           0           6           0           1           0           0   \n",
      "3           0           0           0           0           0           0   \n",
      "4           0           0           0           1           0           0   \n",
      "\n",
      "   GSM1234741  GSM1234742  GSM1234743  GSM1234744  \n",
      "0           0           0           0           0  \n",
      "1           0           0           1           0  \n",
      "2           2           0           0           3  \n",
      "3           0           0           0           0  \n",
      "4           0           0           0           1  \n",
      "\n",
      "[5 rows x 5560 columns]\n"
     ]
    }
   ],
   "source": [
    "dee2 = pd.concat([dee2_genes, dee2_exp], axis=1)\n",
    "dee2.to_csv('dee.csv')   \n",
    "print(dee2.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6bb19de-bd7a-41cd-8580-1beeb9fdaaa9",
   "metadata": {},
   "source": [
    "#### Проведем сортировку"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0af6d820-9f40-4785-bca1-d193d10ba0ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32544\n",
      "52636\n"
     ]
    }
   ],
   "source": [
    "exp = pd.read_csv('expression6.csv', encoding=\"utf8\") # большая таблица\n",
    "dee = pd.read_csv('dee.csv') #маленькая таблица\n",
    "exp.pop('Unnamed: 0')\n",
    "dee.pop('Unnamed: 0')\n",
    "exp=exp.sort_values(by=['ensemble','symbol'])\n",
    "dee=dee.sort_values(by=['ensemble','symbol'])\n",
    "print(exp.shape[0])\n",
    "print(dee.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "02c46506-9070-447f-a638-c101040e9130",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32544\n",
      "9412\n",
      "52636\n",
      "5560\n"
     ]
    }
   ],
   "source": [
    "print(exp.shape[0])\n",
    "print(exp.shape[1])\n",
    "print(dee.shape[0])\n",
    "print(dee.shape[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d3b3690-fc56-4d6e-a99c-9c7948414662",
   "metadata": {},
   "source": [
    "#### Из большой таблицы выделить ту часть, что соответствует маленькой"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ec7a47d-235c-426d-bcf1-fcb3c8ab77ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 ensemble symbol  GSM4042993  GSM4042994  GSM4042995  \\\n",
      "11893  ENSMUSG00000000001  Gnai3        3962        3807        2487   \n",
      "8469   ENSMUSG00000000003   Pbsn           0           0           0   \n",
      "46315  ENSMUSG00000000028  Cdc45        1254        1209        1144   \n",
      "9726   ENSMUSG00000000037  Scml2           0           0           1   \n",
      "39400  ENSMUSG00000000049   Apoh           0           0           3   \n",
      "\n",
      "       GSM4042996  GSM4042997  GSM4042998  GSM4042999  GSM4043000  ...  \\\n",
      "11893        3122        3196        2420        5688        7300  ...   \n",
      "8469            0           0           0           0           0  ...   \n",
      "46315        1362        1310        1264        1109        1569  ...   \n",
      "9726            1           0           0           2           0  ...   \n",
      "39400           0           1           4           0           0  ...   \n",
      "\n",
      "       GSM1234735  GSM1234736  GSM1234737  GSM1234738  GSM1234739  GSM1234740  \\\n",
      "11893        4077        4671        7293        4627        5002        6381   \n",
      "8469            0           0           0           0           0           0   \n",
      "46315         331         302         495         842        1317        2981   \n",
      "9726            2           4           8           5           7          21   \n",
      "39400           7           3           5           5           3           2   \n",
      "\n",
      "       GSM1234741  GSM1234742  GSM1234743  GSM1234744  \n",
      "11893        8072        4725        3897        3205  \n",
      "8469            0           0           0           0  \n",
      "46315        3138        1849        2530        1409  \n",
      "9726           15           0           1           1  \n",
      "39400           7           1          10           0  \n",
      "\n",
      "[5 rows x 5560 columns]\n",
      "30608\n"
     ]
    }
   ],
   "source": [
    "dee_cross=dee[dee['ensemble'].isin(exp['ensemble'])]\n",
    "print(dee_cross.head())\n",
    "print(dee_cross.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e9d85ad8-61c1-4f34-85d2-3d2f731ede9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 ensemble symbol  GSM1356226  GSM1849922  GSM1897413  \\\n",
      "16843  ENSMUSG00000000001  Gnai3       17328        8597         507   \n",
      "23599  ENSMUSG00000000003   Pbsn           0           0           0   \n",
      "3402   ENSMUSG00000000028  Cdc45        1577        5479          27   \n",
      "26959  ENSMUSG00000000037  Scml2          16          86           0   \n",
      "1743   ENSMUSG00000000049   Apoh           1           0           0   \n",
      "\n",
      "       GSM1262088  GSM1119597  GSM1533925  GSM1511235  GSM1808311  ...  \\\n",
      "16843        6175         536        1079        7421       10601  ...   \n",
      "23599           0           0           0           0           0  ...   \n",
      "3402           53         293          14         187         512  ...   \n",
      "26959           7          50           0          10          32  ...   \n",
      "1743            0           0           0           0           0  ...   \n",
      "\n",
      "       GSM5550906  GSM5550908  GSM5550909  GSM5558220  GSM5558221  GSM5558224  \\\n",
      "16843       23609       13185       25213        2788        4192        3695   \n",
      "23599           0           0           0           0           0           0   \n",
      "3402         2540        1260         640         304         275         230   \n",
      "26959           7           0           0           0          10           9   \n",
      "1743            0           0           0           0           0           0   \n",
      "\n",
      "       GSM5558226  GSM5560174  GSM5560175  GSM5560176  \n",
      "16843        3538        6452        7118        6021  \n",
      "23599           0           0           0           0  \n",
      "3402          210        1156         947         399  \n",
      "26959           5          10           9          28  \n",
      "1743            3          11          14           0  \n",
      "\n",
      "[5 rows x 9412 columns]\n",
      "30608\n"
     ]
    }
   ],
   "source": [
    "exp_cross=exp[exp.ensemble.isin(dee_cross.ensemble)  ]\n",
    "print(exp_cross.head())\n",
    "print(exp_cross.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bb7a91a-ed59-42fa-b98a-f4dc054aa0b5",
   "metadata": {},
   "source": [
    "1. Отсортировать положение столбцов\n",
    "2. Перекрестить таблицы по списку столбцов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e37ec24-8442-46b3-8f5c-ad9371647281",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GSM1026888  GSM1026889  GSM1026890  GSM1026891  GSM1026892  GSM1026893  \\\n",
      "16843        5567        5857        4467        5325        6059        5579   \n",
      "23599           0           0           0           0           0           0   \n",
      "3402           16          61          12          85         124         172   \n",
      "26959           2           6           5           0           3          12   \n",
      "1743            2           0           1           1           0           0   \n",
      "\n",
      "       GSM1026894  GSM1026895  GSM1026896  GSM1026897  ...  GSM994541  \\\n",
      "16843        4753        5581        4557        3708  ...       9219   \n",
      "23599           0           0           0           0  ...          0   \n",
      "3402          117         188           7          29  ...       1913   \n",
      "26959          10          16           1           3  ...         26   \n",
      "1743            2           0           3           1  ...          0   \n",
      "\n",
      "       GSM994542  GSM994543  GSM994544  GSM994545  GSM994546  GSM994547  \\\n",
      "16843       4756       7397       4716       6923       6574       8203   \n",
      "23599          0          0          1          0          0          0   \n",
      "3402        1219       1720       1271       1151       1475       1144   \n",
      "26959         30         19         27         24         37         31   \n",
      "1743           1          1          0          0          2          2   \n",
      "\n",
      "       GSM994548            ensemble  symbol  \n",
      "16843       6614  ENSMUSG00000000001   Gnai3  \n",
      "23599          0  ENSMUSG00000000003    Pbsn  \n",
      "3402        1380  ENSMUSG00000000028   Cdc45  \n",
      "26959         23  ENSMUSG00000000037   Scml2  \n",
      "1743           0  ENSMUSG00000000049    Apoh  \n",
      "\n",
      "[5 rows x 9412 columns]\n",
      "       GSM1005492  GSM1026888  GSM1026889  GSM1026890  GSM1026891  GSM1026892  \\\n",
      "11893          94        1857        1946        1460        1731        2064   \n",
      "8469            0           0           0           0           0           0   \n",
      "46315           9           6          17           4          27          46   \n",
      "9726            1           0           1           1           0           1   \n",
      "39400           0           1           0           0           0           0   \n",
      "\n",
      "       GSM1026893  GSM1026894  GSM1026895  GSM1026896  ...  GSM951562  \\\n",
      "11893        1829        1524        1851        1452  ...       1238   \n",
      "8469            0           0           0           0  ...          0   \n",
      "46315          61          43          58           1  ...        126   \n",
      "9726            4           3           3           1  ...          5   \n",
      "39400           0           1           0           0  ...          0   \n",
      "\n",
      "       GSM951563  GSM951564  GSM951565  GSM951566  GSM951567  GSM951568  \\\n",
      "11893       1618       2624       2351       3018       6720       3657   \n",
      "8469           0          0          0          0          0          0   \n",
      "46315         37        281        295        206        146         43   \n",
      "9726           3         12         15          9         10          3   \n",
      "39400          0          0          0          0          0          0   \n",
      "\n",
      "       GSM988446            ensemble  symbol  \n",
      "11893        697  ENSMUSG00000000001   Gnai3  \n",
      "8469           0  ENSMUSG00000000003    Pbsn  \n",
      "46315         76  ENSMUSG00000000028   Cdc45  \n",
      "9726         102  ENSMUSG00000000037   Scml2  \n",
      "39400          0  ENSMUSG00000000049    Apoh  \n",
      "\n",
      "[5 rows x 5560 columns]\n"
     ]
    }
   ],
   "source": [
    "exp_cross = exp_cross[sorted(exp_cross.columns )]\n",
    "dee_cross = dee_cross[sorted(dee_cross.columns )]\n",
    "print(exp_cross.head())\n",
    "print(dee_cross.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "86b983ce-2a10-4dac-9dd2-907e6289afed",
   "metadata": {},
   "outputs": [],
   "source": [
    "col=dee_cross.columns\n",
    "g=[]\n",
    "for i in col:\n",
    "    g.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "366e05b9-1e6b-42be-b272-b8345a57a060",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_cross=exp_cross.loc[:, exp_cross.columns.intersection(g)]\n",
    "col=exp_cross.columns\n",
    "h=[]\n",
    "for i in col:\n",
    "    h.append(i)\n",
    "dee_cross=dee_cross.loc[:, dee_cross.columns.intersection(h)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27a2b9da-cbb7-4213-9cec-19374d22364b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       GSM1026888  GSM1026889  GSM1026890  GSM1026891  GSM1026892  GSM1026893  \\\n",
      "16843        5567        5857        4467        5325        6059        5579   \n",
      "23599           0           0           0           0           0           0   \n",
      "3402           16          61          12          85         124         172   \n",
      "26959           2           6           5           0           3          12   \n",
      "1743            2           0           1           1           0           0   \n",
      "\n",
      "       GSM1026894  GSM1026895  GSM1026896  GSM1026897  ...  GSM951562  \\\n",
      "16843        4753        5581        4557        3708  ...       1090   \n",
      "23599           0           0           0           0  ...          0   \n",
      "3402          117         188           7          29  ...        123   \n",
      "26959          10          16           1           3  ...          4   \n",
      "1743            2           0           3           1  ...          0   \n",
      "\n",
      "       GSM951563  GSM951564  GSM951565  GSM951566  GSM951567  GSM951568  \\\n",
      "16843       1446       2473       2214       2912       6433       3521   \n",
      "23599          0          0          0          0          0          0   \n",
      "3402          33        266        274        196        138         41   \n",
      "26959          4         14         17         13         11          5   \n",
      "1743           0          0          0          0          1          0   \n",
      "\n",
      "       GSM988446            ensemble  symbol  \n",
      "16843          0  ENSMUSG00000000001   Gnai3  \n",
      "23599          4  ENSMUSG00000000003    Pbsn  \n",
      "3402           0  ENSMUSG00000000028   Cdc45  \n",
      "26959         11  ENSMUSG00000000037   Scml2  \n",
      "1743           3  ENSMUSG00000000049    Apoh  \n",
      "\n",
      "[5 rows x 5181 columns]\n",
      "       GSM1026888  GSM1026889  GSM1026890  GSM1026891  GSM1026892  GSM1026893  \\\n",
      "11893        1857        1946        1460        1731        2064        1829   \n",
      "8469            0           0           0           0           0           0   \n",
      "46315           6          17           4          27          46          61   \n",
      "9726            0           1           1           0           1           4   \n",
      "39400           1           0           0           0           0           0   \n",
      "\n",
      "       GSM1026894  GSM1026895  GSM1026896  GSM1026897  ...  GSM951562  \\\n",
      "11893        1524        1851        1452        1211  ...       1238   \n",
      "8469            0           0           0           0  ...          0   \n",
      "46315          43          58           1           7  ...        126   \n",
      "9726            3           3           1           0  ...          5   \n",
      "39400           1           0           0           0  ...          0   \n",
      "\n",
      "       GSM951563  GSM951564  GSM951565  GSM951566  GSM951567  GSM951568  \\\n",
      "11893       1618       2624       2351       3018       6720       3657   \n",
      "8469           0          0          0          0          0          0   \n",
      "46315         37        281        295        206        146         43   \n",
      "9726           3         12         15          9         10          3   \n",
      "39400          0          0          0          0          0          0   \n",
      "\n",
      "       GSM988446            ensemble  symbol  \n",
      "11893        697  ENSMUSG00000000001   Gnai3  \n",
      "8469           0  ENSMUSG00000000003    Pbsn  \n",
      "46315         76  ENSMUSG00000000028   Cdc45  \n",
      "9726         102  ENSMUSG00000000037   Scml2  \n",
      "39400          0  ENSMUSG00000000049    Apoh  \n",
      "\n",
      "[5 rows x 5181 columns]\n"
     ]
    }
   ],
   "source": [
    "print(exp_cross.head())\n",
    "print(dee_cross.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "763a05c4-4e69-49cb-bb9d-f99ca7d43ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_cross.to_csv('exp_cross.csv')\n",
    "dee_cross.to_csv('dee_cross.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba90f86-2c09-44fd-90d1-4ec8dd5a8331",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
